{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d941f744",
   "metadata": {},
   "source": [
    "# Overview: checking metadata and samplesheet for all RUNS"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da652c79",
   "metadata": {},
   "source": [
    "## Preparation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4510d8e5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All available runs: R7288, R7297, R7312, R7331, R7347, R7353, R7373, R7374, R7393, R7400\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd \n",
    "import numpy as np \n",
    "import matplotlib.pyplot as plt \n",
    "import seaborn as sns \n",
    "import os\n",
    "import pathlib\n",
    "import pysam\n",
    "import pyfaidx\n",
    "import warnings\n",
    "import re\n",
    "import argparse\n",
    "import sys\n",
    "import subprocess\n",
    "from tqdm import tqdm \n",
    "\n",
    "tqdm.pandas()\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "main_outptudir = \"/workdir/outdir/pipeline_output\"\n",
    "umt_distance = 0\n",
    "analysis_output_version = \"20250614\"\n",
    "name = \"Vi\"\n",
    "\n",
    "path_to_main_output = os.path.join(main_outptudir, \"data_analysis\", name, analysis_output_version)\n",
    "path_to_01_output = os.path.join(path_to_main_output, \"01_output\")\n",
    "os.system(f\"mkdir -p {path_to_01_output}\")\n",
    "\n",
    "rundf = pd.read_excel(\"../panel_design/All_panel_designs_20250601.xlsx\", sheet_name=\"Runs_tracking\")\n",
    "rundf.columns = [\"Run\", \"SampleID\", \"UMI\", \"Panel_version\", \"Name\", \"amplicon_name\", \"serial\"]\n",
    "\n",
    "umi_rundf = rundf[(rundf[\"UMI\"] == \"Yes\") & (rundf[\"Name\"] == name) ]\n",
    "\n",
    "collect_all_sample_sheets = [item for item in pathlib.Path(\"../experiments/\").glob(\"SampleSheet_batch_*.csv\")]\n",
    "\n",
    "samplesheet = pd.DataFrame()\n",
    "for file in collect_all_sample_sheets:\n",
    "    df = pd.read_csv(file, sep=\",\")\n",
    "    df[\"batch\"] = file.name.replace(\".csv\", \"\")\n",
    "    samplesheet = pd.concat([samplesheet, df], ignore_index=True)\n",
    "\n",
    "samplesheet.columns = [\"filename\", \"FASTQ1\", \"FASTQ2\", \"batch\"]\n",
    "samplesheet[\"Run\"] = samplesheet[\"FASTQ1\"].apply(lambda x: x.split(\"/\")[5] if x.split(\"/\")[5] != \"2025\" else x.split(\"/\")[6])\n",
    "samplesheet[\"SampleID\"] = samplesheet[\"filename\"].apply(lambda x: x.split(\"_\")[0].split(\"-\")[1])\n",
    "print(f\"All available runs: {\", \".join(sorted(samplesheet['Run'].unique()))}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db80ab64",
   "metadata": {},
   "source": [
    "## Check if there is any missing FASTQ samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f8de7362",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Run</th>\n",
       "      <th>SampleID</th>\n",
       "      <th>UMI</th>\n",
       "      <th>Panel_version</th>\n",
       "      <th>Name</th>\n",
       "      <th>amplicon_name</th>\n",
       "      <th>serial</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>757</th>\n",
       "      <td>R7434</td>\n",
       "      <td>TMPC4S1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>CRC_panel version 1.2</td>\n",
       "      <td>Vi</td>\n",
       "      <td>H1, H3</td>\n",
       "      <td>S1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>758</th>\n",
       "      <td>R7434</td>\n",
       "      <td>TMPC4S2</td>\n",
       "      <td>Yes</td>\n",
       "      <td>CRC_panel version 1.2</td>\n",
       "      <td>Vi</td>\n",
       "      <td>H1, H3</td>\n",
       "      <td>S2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>759</th>\n",
       "      <td>R7434</td>\n",
       "      <td>TMPC4S3</td>\n",
       "      <td>Yes</td>\n",
       "      <td>CRC_panel version 1.2</td>\n",
       "      <td>Vi</td>\n",
       "      <td>H1, H3</td>\n",
       "      <td>S3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>761</th>\n",
       "      <td>R7434</td>\n",
       "      <td>TMPC4S5</td>\n",
       "      <td>Yes</td>\n",
       "      <td>CRC_panel version 1.2</td>\n",
       "      <td>Vi</td>\n",
       "      <td>H1, H3</td>\n",
       "      <td>S5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>762</th>\n",
       "      <td>R7434</td>\n",
       "      <td>TMPC4S6</td>\n",
       "      <td>Yes</td>\n",
       "      <td>CRC_panel version 1.2</td>\n",
       "      <td>Vi</td>\n",
       "      <td>H1, H3</td>\n",
       "      <td>S6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>763</th>\n",
       "      <td>R7434</td>\n",
       "      <td>TMPC4S7</td>\n",
       "      <td>Yes</td>\n",
       "      <td>CRC_panel version 1.2</td>\n",
       "      <td>Vi</td>\n",
       "      <td>H1, H3</td>\n",
       "      <td>S7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>764</th>\n",
       "      <td>R7434</td>\n",
       "      <td>TMPL7S1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Lung_panel version 1.1</td>\n",
       "      <td>Vi</td>\n",
       "      <td>L1, L2, L3</td>\n",
       "      <td>S1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>765</th>\n",
       "      <td>R7434</td>\n",
       "      <td>TMPL7S2</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Lung_panel version 1.1</td>\n",
       "      <td>Vi</td>\n",
       "      <td>L1, L2, L3</td>\n",
       "      <td>S2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>766</th>\n",
       "      <td>R7434</td>\n",
       "      <td>TMPL7S3</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Lung_panel version 1.1</td>\n",
       "      <td>Vi</td>\n",
       "      <td>L1, L2, L3</td>\n",
       "      <td>S3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>767</th>\n",
       "      <td>R7434</td>\n",
       "      <td>TMPL7S4</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Lung_panel version 1.1</td>\n",
       "      <td>Vi</td>\n",
       "      <td>L1, L2, L3</td>\n",
       "      <td>S4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>768</th>\n",
       "      <td>R7434</td>\n",
       "      <td>TMPL7S5</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Lung_panel version 1.1</td>\n",
       "      <td>Vi</td>\n",
       "      <td>L1, L2, L3</td>\n",
       "      <td>S5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>769</th>\n",
       "      <td>R7434</td>\n",
       "      <td>TMPL7S6</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Lung_panel version 1.1</td>\n",
       "      <td>Vi</td>\n",
       "      <td>L1, L2, L3</td>\n",
       "      <td>S6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>770</th>\n",
       "      <td>R7434</td>\n",
       "      <td>TMPL7S7</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Lung_panel version 1.1</td>\n",
       "      <td>Vi</td>\n",
       "      <td>L1, L2, L3</td>\n",
       "      <td>S7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771</th>\n",
       "      <td>R7434</td>\n",
       "      <td>TMCONTROL146</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Lung_panel version 1.1</td>\n",
       "      <td>Vi</td>\n",
       "      <td>L1, L2, L3</td>\n",
       "      <td>cfDNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Run      SampleID  UMI           Panel_version Name amplicon_name  \\\n",
       "757  R7434       TMPC4S1  Yes   CRC_panel version 1.2   Vi        H1, H3   \n",
       "758  R7434       TMPC4S2  Yes   CRC_panel version 1.2   Vi        H1, H3   \n",
       "759  R7434       TMPC4S3  Yes   CRC_panel version 1.2   Vi        H1, H3   \n",
       "761  R7434       TMPC4S5  Yes   CRC_panel version 1.2   Vi        H1, H3   \n",
       "762  R7434       TMPC4S6  Yes   CRC_panel version 1.2   Vi        H1, H3   \n",
       "763  R7434       TMPC4S7  Yes   CRC_panel version 1.2   Vi        H1, H3   \n",
       "764  R7434       TMPL7S1  Yes  Lung_panel version 1.1   Vi    L1, L2, L3   \n",
       "765  R7434       TMPL7S2  Yes  Lung_panel version 1.1   Vi    L1, L2, L3   \n",
       "766  R7434       TMPL7S3  Yes  Lung_panel version 1.1   Vi    L1, L2, L3   \n",
       "767  R7434       TMPL7S4  Yes  Lung_panel version 1.1   Vi    L1, L2, L3   \n",
       "768  R7434       TMPL7S5  Yes  Lung_panel version 1.1   Vi    L1, L2, L3   \n",
       "769  R7434       TMPL7S6  Yes  Lung_panel version 1.1   Vi    L1, L2, L3   \n",
       "770  R7434       TMPL7S7  Yes  Lung_panel version 1.1   Vi    L1, L2, L3   \n",
       "771  R7434  TMCONTROL146  Yes  Lung_panel version 1.1   Vi    L1, L2, L3   \n",
       "\n",
       "    serial  \n",
       "757     S1  \n",
       "758     S2  \n",
       "759     S3  \n",
       "761     S5  \n",
       "762     S6  \n",
       "763     S7  \n",
       "764     S1  \n",
       "765     S2  \n",
       "766     S3  \n",
       "767     S4  \n",
       "768     S5  \n",
       "769     S6  \n",
       "770     S7  \n",
       "771  cfDNA  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "missing_fastq = [item for item in umi_rundf.SampleID.unique() if item not in samplesheet.SampleID.unique()]\n",
    "\n",
    "umi_rundf[umi_rundf.SampleID.isin(missing_fastq)]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96bb5f1b",
   "metadata": {},
   "source": [
    "## Match output paths to main metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0fe5891c",
   "metadata": {},
   "source": [
    "since there is duplicated SampleID, we should merge labcode and Run together to make them unique"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "226d925b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "reading in saved processed UMI runs tracking file...\n"
     ]
    }
   ],
   "source": [
    "if os.path.isfile(os.path.join(path_to_01_output, \"UMI_runs_tracking.xlsx\")) == False:\n",
    "    umi_rundf[\"uniqueID\"] = umi_rundf[[\"SampleID\", \"Run\"]].apply(lambda x: f\"{x[0]}_{x[1]}\", axis=1)\n",
    "    samplesheet[\"uniqueID\"] = samplesheet[[\"SampleID\", \"Run\"]].apply(lambda x: f\"{x[0]}_{x[1]}\", axis=1)\n",
    "\n",
    "    final_umi_rundf = umi_rundf.merge(samplesheet.drop([\"SampleID\", \"Run\"], axis = 1), left_on = \"uniqueID\", right_on = \"uniqueID\")\n",
    "    final_umi_rundf[\"UMI_cov_file\"] = final_umi_rundf[[\"batch\", \"filename\"]].apply(lambda x: os.path.join(\n",
    "        main_outptudir, x[0], f\"UMT_DISTANCE_{umt_distance}\", \"BISMARK_COV\", f\"{x[1]}.connor_R1_bismark_bt2_pe.bedGraph.gz.bismark.zero.cov\"\n",
    "    ), axis = 1)\n",
    "    final_umi_rundf[\"nonUMI_cov_file\"] = final_umi_rundf[[\"batch\", \"filename\"]].apply(lambda x: os.path.join(\n",
    "        main_outptudir, x[0], f\"UMT_DISTANCE_{umt_distance}\", \"BISMARK_COV\", f\"{x[1]}.CutAdapt_R1_bismark_bt2_pe.bedGraph.gz.bismark.zero.cov\"\n",
    "    ), axis = 1)\n",
    "\n",
    "    final_umi_rundf[\"check_UMI\"] = final_umi_rundf[\"UMI_cov_file\"].apply(lambda x: os.path.exists(x))\n",
    "    final_umi_rundf[\"check_nonUMI\"] = final_umi_rundf[\"nonUMI_cov_file\"].apply(lambda x: os.path.exists(x))\n",
    "\n",
    "    final_umi_rundf[\"unmapped_bam\"] = final_umi_rundf[[\"batch\", \"filename\"]].apply(\n",
    "        lambda x: os.path.join(main_outptudir, x[0], f\"UMT_DISTANCE_{umt_distance}\", \"BISMARK_ALIGNMENT_UNMAPPED_BAM\", f\"{x[1]}_R1.UMIprocessed_bismark_bt2_pe.bam\"), axis = 1\n",
    "    )\n",
    "    final_umi_rundf[\"connor_annotated\"] = final_umi_rundf[[\"batch\", \"filename\"]].apply(\n",
    "        lambda x: os.path.join(main_outptudir, x[0], f\"UMT_DISTANCE_{umt_distance}\", \"CONNOR_UMI_OUTPUT\", f\"{x[1]}.connor.fully_annotated.bam\"), axis = 1\n",
    "    )\n",
    "    final_umi_rundf[\"connor_fastq\"] = final_umi_rundf[[\"batch\", \"filename\"]].apply(\n",
    "        lambda x: os.path.join(main_outptudir, x[0], f\"UMT_DISTANCE_{umt_distance}\", \"CONNOR_UMI_OUTPUT\", f\"{x[1]}.connor_R1.fastq.gz\"), axis = 1\n",
    "    )\n",
    "    final_umi_rundf[\"umi_bam\"] = final_umi_rundf[[\"batch\", \"filename\"]].apply(\n",
    "        lambda x: os.path.join(main_outptudir, x[0], f\"UMT_DISTANCE_{umt_distance}\", \"BISMARK_ALIGNMENT\", f\"{x[1]}.connor_R1_bismark_bt2_pe.sorted.bam\"), axis = 1\n",
    "    )\n",
    "    final_umi_rundf[\"non_umi_bam\"] = final_umi_rundf[[\"batch\", \"filename\"]].apply(\n",
    "        lambda x: os.path.join(main_outptudir, x[0], f\"UMT_DISTANCE_{umt_distance}\", \"BISMARK_ALIGNMENT\", f\"{x[1]}.CutAdapt_R1_bismark_bt2_pe.sorted.bam\"), axis = 1\n",
    "    )\n",
    "\n",
    "    tqdm.pandas()\n",
    "    for col in [\"unmapped_bam\", \"connor_annotated\", \"connor_fastq\", \"umi_bam\", \"non_umi_bam\" , \"UMI_cov_file\"]:\n",
    "        print(f\"working on column: {col}\")\n",
    "        def _get_line_count(x):\n",
    "            check_file = os.path.exists(x)\n",
    "            x_raw = x.replace(\".sorted.bam\", \".bam\")\n",
    "            if check_file == False:\n",
    "                if os.path.exists(x_raw) == True: \n",
    "                    # print(f\"sorted bam file is not available, but raw bam file is available, sorting and indexing it now...\")\n",
    "                    os.system(f\"samtools sort -@ 15 {x_raw} -o {x}\")\n",
    "                    os.system(f\"samtools index {x}\")\n",
    "                else:\n",
    "                    # rint(f\"File {x} does not exist, skipping line count for this file.\")\n",
    "                    return 0\n",
    "            if \".bam\" in x:\n",
    "                result = subprocess.run(f\"samtools view {x} -c\", shell=True, capture_output=True, text=True)\n",
    "            elif \".fastq.gz\" in x:\n",
    "                result = subprocess.run(f\"zcat {x} | wc -l\", shell=True, capture_output=True, text=True)\n",
    "            else:\n",
    "                result = subprocess.run(f\"cat {x} | wc -l\", shell=True, capture_output=True, text=True)\n",
    "            return int(result.stdout.strip())\n",
    "            \n",
    "        final_umi_rundf[f\"count_{col}\"] = final_umi_rundf[col].progress_apply(_get_line_count)    \n",
    "\n",
    "    final_umi_rundf.to_excel(os.path.join(path_to_01_output, \"UMI_runs_tracking.xlsx\"), index=False)\n",
    "else:\n",
    "    print(\"reading in saved processed UMI runs tracking file...\")\n",
    "    final_umi_rundf = pd.read_excel(os.path.join(path_to_01_output, \"UMI_runs_tracking.xlsx\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "530f348d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Run</th>\n",
       "      <th>SampleID</th>\n",
       "      <th>UMI</th>\n",
       "      <th>Panel_version</th>\n",
       "      <th>Name</th>\n",
       "      <th>amplicon_name</th>\n",
       "      <th>serial</th>\n",
       "      <th>uniqueID</th>\n",
       "      <th>filename</th>\n",
       "      <th>FASTQ1</th>\n",
       "      <th>...</th>\n",
       "      <th>connor_annotated</th>\n",
       "      <th>connor_fastq</th>\n",
       "      <th>umi_bam</th>\n",
       "      <th>non_umi_bam</th>\n",
       "      <th>count_unmapped_bam</th>\n",
       "      <th>count_connor_annotated</th>\n",
       "      <th>count_connor_fastq</th>\n",
       "      <th>count_umi_bam</th>\n",
       "      <th>count_non_umi_bam</th>\n",
       "      <th>count_UMI_cov_file</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>R7288</td>\n",
       "      <td>TML1S1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Lung_panel version 1.1</td>\n",
       "      <td>Vi</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S1</td>\n",
       "      <td>TML1S1_R7288</td>\n",
       "      <td>1-TML1S1_S7501-S7701</td>\n",
       "      <td>/mnt/GS-BACKUP05/FASTQ/2025/R7288/zatm/1-TML1S...</td>\n",
       "      <td>...</td>\n",
       "      <td>/workdir/outdir/pipeline_output/SampleSheet_ba...</td>\n",
       "      <td>/workdir/outdir/pipeline_output/SampleSheet_ba...</td>\n",
       "      <td>/workdir/outdir/pipeline_output/SampleSheet_ba...</td>\n",
       "      <td>/workdir/outdir/pipeline_output/SampleSheet_ba...</td>\n",
       "      <td>2645274</td>\n",
       "      <td>2639164</td>\n",
       "      <td>34584</td>\n",
       "      <td>340</td>\n",
       "      <td>2645274</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>R7288</td>\n",
       "      <td>TML1S2</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Lung_panel version 1.1</td>\n",
       "      <td>Vi</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S2</td>\n",
       "      <td>TML1S2_R7288</td>\n",
       "      <td>2-TML1S2_S7502-S7702</td>\n",
       "      <td>/mnt/GS-BACKUP05/FASTQ/2025/R7288/zatm/2-TML1S...</td>\n",
       "      <td>...</td>\n",
       "      <td>/workdir/outdir/pipeline_output/SampleSheet_ba...</td>\n",
       "      <td>/workdir/outdir/pipeline_output/SampleSheet_ba...</td>\n",
       "      <td>/workdir/outdir/pipeline_output/SampleSheet_ba...</td>\n",
       "      <td>/workdir/outdir/pipeline_output/SampleSheet_ba...</td>\n",
       "      <td>2195246</td>\n",
       "      <td>2181957</td>\n",
       "      <td>32044</td>\n",
       "      <td>466</td>\n",
       "      <td>2195246</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>R7288</td>\n",
       "      <td>TML1S3</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Lung_panel version 1.1</td>\n",
       "      <td>Vi</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S3</td>\n",
       "      <td>TML1S3_R7288</td>\n",
       "      <td>3-TML1S3_S7503-S7703</td>\n",
       "      <td>/mnt/GS-BACKUP05/FASTQ/2025/R7288/zatm/3-TML1S...</td>\n",
       "      <td>...</td>\n",
       "      <td>/workdir/outdir/pipeline_output/SampleSheet_ba...</td>\n",
       "      <td>/workdir/outdir/pipeline_output/SampleSheet_ba...</td>\n",
       "      <td>/workdir/outdir/pipeline_output/SampleSheet_ba...</td>\n",
       "      <td>/workdir/outdir/pipeline_output/SampleSheet_ba...</td>\n",
       "      <td>3528722</td>\n",
       "      <td>3519234</td>\n",
       "      <td>38204</td>\n",
       "      <td>542</td>\n",
       "      <td>3528722</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>R7288</td>\n",
       "      <td>TML1S4</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Lung_panel version 1.1</td>\n",
       "      <td>Vi</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S4</td>\n",
       "      <td>TML1S4_R7288</td>\n",
       "      <td>4-TML1S4_S7504-S7704</td>\n",
       "      <td>/mnt/GS-BACKUP05/FASTQ/2025/R7288/zatm/4-TML1S...</td>\n",
       "      <td>...</td>\n",
       "      <td>/workdir/outdir/pipeline_output/SampleSheet_ba...</td>\n",
       "      <td>/workdir/outdir/pipeline_output/SampleSheet_ba...</td>\n",
       "      <td>/workdir/outdir/pipeline_output/SampleSheet_ba...</td>\n",
       "      <td>/workdir/outdir/pipeline_output/SampleSheet_ba...</td>\n",
       "      <td>3210054</td>\n",
       "      <td>3200836</td>\n",
       "      <td>36748</td>\n",
       "      <td>498</td>\n",
       "      <td>3210054</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>R7288</td>\n",
       "      <td>TML1S5</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Lung_panel version 1.1</td>\n",
       "      <td>Vi</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S5</td>\n",
       "      <td>TML1S5_R7288</td>\n",
       "      <td>5-TML1S5_S7505-S7705</td>\n",
       "      <td>/mnt/GS-BACKUP05/FASTQ/2025/R7288/zatm/5-TML1S...</td>\n",
       "      <td>...</td>\n",
       "      <td>/workdir/outdir/pipeline_output/SampleSheet_ba...</td>\n",
       "      <td>/workdir/outdir/pipeline_output/SampleSheet_ba...</td>\n",
       "      <td>/workdir/outdir/pipeline_output/SampleSheet_ba...</td>\n",
       "      <td>/workdir/outdir/pipeline_output/SampleSheet_ba...</td>\n",
       "      <td>3048870</td>\n",
       "      <td>3039512</td>\n",
       "      <td>36012</td>\n",
       "      <td>596</td>\n",
       "      <td>3048870</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>309</th>\n",
       "      <td>R7400</td>\n",
       "      <td>TMPL6S3I3</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Lung_panel version 1.1</td>\n",
       "      <td>Vi</td>\n",
       "      <td>L3, L6, L12</td>\n",
       "      <td>S3</td>\n",
       "      <td>TMPL6S3I3_R7400</td>\n",
       "      <td>10-TMPL6S3I3_S7590-S7790</td>\n",
       "      <td>/mnt/GS-BACKUP05/FASTQ/2025/R7400/zatm/10-TMPL...</td>\n",
       "      <td>...</td>\n",
       "      <td>/workdir/outdir/pipeline_output/SampleSheet_ba...</td>\n",
       "      <td>/workdir/outdir/pipeline_output/SampleSheet_ba...</td>\n",
       "      <td>/workdir/outdir/pipeline_output/SampleSheet_ba...</td>\n",
       "      <td>/workdir/outdir/pipeline_output/SampleSheet_ba...</td>\n",
       "      <td>191490</td>\n",
       "      <td>189409</td>\n",
       "      <td>16324</td>\n",
       "      <td>262</td>\n",
       "      <td>191490</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>310</th>\n",
       "      <td>R7400</td>\n",
       "      <td>TMPL6S4I3</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Lung_panel version 1.1</td>\n",
       "      <td>Vi</td>\n",
       "      <td>L3, L6, L12</td>\n",
       "      <td>S4</td>\n",
       "      <td>TMPL6S4I3_R7400</td>\n",
       "      <td>11-TMPL6S4I3_S7591-S7791</td>\n",
       "      <td>/mnt/GS-BACKUP05/FASTQ/2025/R7400/zatm/11-TMPL...</td>\n",
       "      <td>...</td>\n",
       "      <td>/workdir/outdir/pipeline_output/SampleSheet_ba...</td>\n",
       "      <td>/workdir/outdir/pipeline_output/SampleSheet_ba...</td>\n",
       "      <td>/workdir/outdir/pipeline_output/SampleSheet_ba...</td>\n",
       "      <td>/workdir/outdir/pipeline_output/SampleSheet_ba...</td>\n",
       "      <td>363520</td>\n",
       "      <td>359986</td>\n",
       "      <td>17832</td>\n",
       "      <td>166</td>\n",
       "      <td>363520</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311</th>\n",
       "      <td>R7400</td>\n",
       "      <td>TMPL6S5I3</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Lung_panel version 1.1</td>\n",
       "      <td>Vi</td>\n",
       "      <td>L3, L6, L12</td>\n",
       "      <td>S5</td>\n",
       "      <td>TMPL6S5I3_R7400</td>\n",
       "      <td>12-TMPL6S5I3_S7592-S7792</td>\n",
       "      <td>/mnt/GS-BACKUP05/FASTQ/2025/R7400/zatm/12-TMPL...</td>\n",
       "      <td>...</td>\n",
       "      <td>/workdir/outdir/pipeline_output/SampleSheet_ba...</td>\n",
       "      <td>/workdir/outdir/pipeline_output/SampleSheet_ba...</td>\n",
       "      <td>/workdir/outdir/pipeline_output/SampleSheet_ba...</td>\n",
       "      <td>/workdir/outdir/pipeline_output/SampleSheet_ba...</td>\n",
       "      <td>23070</td>\n",
       "      <td>22344</td>\n",
       "      <td>5640</td>\n",
       "      <td>16</td>\n",
       "      <td>23070</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>312</th>\n",
       "      <td>R7400</td>\n",
       "      <td>TMPL6S6I3</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Lung_panel version 1.1</td>\n",
       "      <td>Vi</td>\n",
       "      <td>L3, L6, L12</td>\n",
       "      <td>S6</td>\n",
       "      <td>TMPL6S6I3_R7400</td>\n",
       "      <td>13-TMPL6S6I3_S7593-S7793</td>\n",
       "      <td>/mnt/GS-BACKUP05/FASTQ/2025/R7400/zatm/13-TMPL...</td>\n",
       "      <td>...</td>\n",
       "      <td>/workdir/outdir/pipeline_output/SampleSheet_ba...</td>\n",
       "      <td>/workdir/outdir/pipeline_output/SampleSheet_ba...</td>\n",
       "      <td>/workdir/outdir/pipeline_output/SampleSheet_ba...</td>\n",
       "      <td>/workdir/outdir/pipeline_output/SampleSheet_ba...</td>\n",
       "      <td>1135294</td>\n",
       "      <td>1113771</td>\n",
       "      <td>33808</td>\n",
       "      <td>352</td>\n",
       "      <td>1135294</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>313</th>\n",
       "      <td>R7400</td>\n",
       "      <td>TMPL6S7I3</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Lung_panel version 1.1</td>\n",
       "      <td>Vi</td>\n",
       "      <td>L3, L6, L12</td>\n",
       "      <td>S7</td>\n",
       "      <td>TMPL6S7I3_R7400</td>\n",
       "      <td>14-TMPL6S7I3_S7594-S7794</td>\n",
       "      <td>/mnt/GS-BACKUP05/FASTQ/2025/R7400/zatm/14-TMPL...</td>\n",
       "      <td>...</td>\n",
       "      <td>/workdir/outdir/pipeline_output/SampleSheet_ba...</td>\n",
       "      <td>/workdir/outdir/pipeline_output/SampleSheet_ba...</td>\n",
       "      <td>/workdir/outdir/pipeline_output/SampleSheet_ba...</td>\n",
       "      <td>/workdir/outdir/pipeline_output/SampleSheet_ba...</td>\n",
       "      <td>1552282</td>\n",
       "      <td>1511620</td>\n",
       "      <td>34396</td>\n",
       "      <td>406</td>\n",
       "      <td>1552282</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>314 rows Ã— 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Run   SampleID  UMI           Panel_version Name amplicon_name serial  \\\n",
       "0    R7288     TML1S1  Yes  Lung_panel version 1.1   Vi           NaN     S1   \n",
       "1    R7288     TML1S2  Yes  Lung_panel version 1.1   Vi           NaN     S2   \n",
       "2    R7288     TML1S3  Yes  Lung_panel version 1.1   Vi           NaN     S3   \n",
       "3    R7288     TML1S4  Yes  Lung_panel version 1.1   Vi           NaN     S4   \n",
       "4    R7288     TML1S5  Yes  Lung_panel version 1.1   Vi           NaN     S5   \n",
       "..     ...        ...  ...                     ...  ...           ...    ...   \n",
       "309  R7400  TMPL6S3I3  Yes  Lung_panel version 1.1   Vi   L3, L6, L12     S3   \n",
       "310  R7400  TMPL6S4I3  Yes  Lung_panel version 1.1   Vi   L3, L6, L12     S4   \n",
       "311  R7400  TMPL6S5I3  Yes  Lung_panel version 1.1   Vi   L3, L6, L12     S5   \n",
       "312  R7400  TMPL6S6I3  Yes  Lung_panel version 1.1   Vi   L3, L6, L12     S6   \n",
       "313  R7400  TMPL6S7I3  Yes  Lung_panel version 1.1   Vi   L3, L6, L12     S7   \n",
       "\n",
       "            uniqueID                  filename  \\\n",
       "0       TML1S1_R7288      1-TML1S1_S7501-S7701   \n",
       "1       TML1S2_R7288      2-TML1S2_S7502-S7702   \n",
       "2       TML1S3_R7288      3-TML1S3_S7503-S7703   \n",
       "3       TML1S4_R7288      4-TML1S4_S7504-S7704   \n",
       "4       TML1S5_R7288      5-TML1S5_S7505-S7705   \n",
       "..               ...                       ...   \n",
       "309  TMPL6S3I3_R7400  10-TMPL6S3I3_S7590-S7790   \n",
       "310  TMPL6S4I3_R7400  11-TMPL6S4I3_S7591-S7791   \n",
       "311  TMPL6S5I3_R7400  12-TMPL6S5I3_S7592-S7792   \n",
       "312  TMPL6S6I3_R7400  13-TMPL6S6I3_S7593-S7793   \n",
       "313  TMPL6S7I3_R7400  14-TMPL6S7I3_S7594-S7794   \n",
       "\n",
       "                                                FASTQ1  ...  \\\n",
       "0    /mnt/GS-BACKUP05/FASTQ/2025/R7288/zatm/1-TML1S...  ...   \n",
       "1    /mnt/GS-BACKUP05/FASTQ/2025/R7288/zatm/2-TML1S...  ...   \n",
       "2    /mnt/GS-BACKUP05/FASTQ/2025/R7288/zatm/3-TML1S...  ...   \n",
       "3    /mnt/GS-BACKUP05/FASTQ/2025/R7288/zatm/4-TML1S...  ...   \n",
       "4    /mnt/GS-BACKUP05/FASTQ/2025/R7288/zatm/5-TML1S...  ...   \n",
       "..                                                 ...  ...   \n",
       "309  /mnt/GS-BACKUP05/FASTQ/2025/R7400/zatm/10-TMPL...  ...   \n",
       "310  /mnt/GS-BACKUP05/FASTQ/2025/R7400/zatm/11-TMPL...  ...   \n",
       "311  /mnt/GS-BACKUP05/FASTQ/2025/R7400/zatm/12-TMPL...  ...   \n",
       "312  /mnt/GS-BACKUP05/FASTQ/2025/R7400/zatm/13-TMPL...  ...   \n",
       "313  /mnt/GS-BACKUP05/FASTQ/2025/R7400/zatm/14-TMPL...  ...   \n",
       "\n",
       "                                      connor_annotated  \\\n",
       "0    /workdir/outdir/pipeline_output/SampleSheet_ba...   \n",
       "1    /workdir/outdir/pipeline_output/SampleSheet_ba...   \n",
       "2    /workdir/outdir/pipeline_output/SampleSheet_ba...   \n",
       "3    /workdir/outdir/pipeline_output/SampleSheet_ba...   \n",
       "4    /workdir/outdir/pipeline_output/SampleSheet_ba...   \n",
       "..                                                 ...   \n",
       "309  /workdir/outdir/pipeline_output/SampleSheet_ba...   \n",
       "310  /workdir/outdir/pipeline_output/SampleSheet_ba...   \n",
       "311  /workdir/outdir/pipeline_output/SampleSheet_ba...   \n",
       "312  /workdir/outdir/pipeline_output/SampleSheet_ba...   \n",
       "313  /workdir/outdir/pipeline_output/SampleSheet_ba...   \n",
       "\n",
       "                                          connor_fastq  \\\n",
       "0    /workdir/outdir/pipeline_output/SampleSheet_ba...   \n",
       "1    /workdir/outdir/pipeline_output/SampleSheet_ba...   \n",
       "2    /workdir/outdir/pipeline_output/SampleSheet_ba...   \n",
       "3    /workdir/outdir/pipeline_output/SampleSheet_ba...   \n",
       "4    /workdir/outdir/pipeline_output/SampleSheet_ba...   \n",
       "..                                                 ...   \n",
       "309  /workdir/outdir/pipeline_output/SampleSheet_ba...   \n",
       "310  /workdir/outdir/pipeline_output/SampleSheet_ba...   \n",
       "311  /workdir/outdir/pipeline_output/SampleSheet_ba...   \n",
       "312  /workdir/outdir/pipeline_output/SampleSheet_ba...   \n",
       "313  /workdir/outdir/pipeline_output/SampleSheet_ba...   \n",
       "\n",
       "                                               umi_bam  \\\n",
       "0    /workdir/outdir/pipeline_output/SampleSheet_ba...   \n",
       "1    /workdir/outdir/pipeline_output/SampleSheet_ba...   \n",
       "2    /workdir/outdir/pipeline_output/SampleSheet_ba...   \n",
       "3    /workdir/outdir/pipeline_output/SampleSheet_ba...   \n",
       "4    /workdir/outdir/pipeline_output/SampleSheet_ba...   \n",
       "..                                                 ...   \n",
       "309  /workdir/outdir/pipeline_output/SampleSheet_ba...   \n",
       "310  /workdir/outdir/pipeline_output/SampleSheet_ba...   \n",
       "311  /workdir/outdir/pipeline_output/SampleSheet_ba...   \n",
       "312  /workdir/outdir/pipeline_output/SampleSheet_ba...   \n",
       "313  /workdir/outdir/pipeline_output/SampleSheet_ba...   \n",
       "\n",
       "                                           non_umi_bam  count_unmapped_bam  \\\n",
       "0    /workdir/outdir/pipeline_output/SampleSheet_ba...             2645274   \n",
       "1    /workdir/outdir/pipeline_output/SampleSheet_ba...             2195246   \n",
       "2    /workdir/outdir/pipeline_output/SampleSheet_ba...             3528722   \n",
       "3    /workdir/outdir/pipeline_output/SampleSheet_ba...             3210054   \n",
       "4    /workdir/outdir/pipeline_output/SampleSheet_ba...             3048870   \n",
       "..                                                 ...                 ...   \n",
       "309  /workdir/outdir/pipeline_output/SampleSheet_ba...              191490   \n",
       "310  /workdir/outdir/pipeline_output/SampleSheet_ba...              363520   \n",
       "311  /workdir/outdir/pipeline_output/SampleSheet_ba...               23070   \n",
       "312  /workdir/outdir/pipeline_output/SampleSheet_ba...             1135294   \n",
       "313  /workdir/outdir/pipeline_output/SampleSheet_ba...             1552282   \n",
       "\n",
       "     count_connor_annotated count_connor_fastq count_umi_bam  \\\n",
       "0                   2639164              34584           340   \n",
       "1                   2181957              32044           466   \n",
       "2                   3519234              38204           542   \n",
       "3                   3200836              36748           498   \n",
       "4                   3039512              36012           596   \n",
       "..                      ...                ...           ...   \n",
       "309                  189409              16324           262   \n",
       "310                  359986              17832           166   \n",
       "311                   22344               5640            16   \n",
       "312                 1113771              33808           352   \n",
       "313                 1511620              34396           406   \n",
       "\n",
       "    count_non_umi_bam count_UMI_cov_file  \n",
       "0             2645274                  4  \n",
       "1             2195246                  4  \n",
       "2             3528722                  4  \n",
       "3             3210054                  4  \n",
       "4             3048870                  4  \n",
       "..                ...                ...  \n",
       "309            191490                 13  \n",
       "310            363520                 35  \n",
       "311             23070                 15  \n",
       "312           1135294                 35  \n",
       "313           1552282                 35  \n",
       "\n",
       "[314 rows x 27 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_umi_rundf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b8f0dec",
   "metadata": {},
   "source": [
    "# Helper functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "9ccb84ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_refseq(path_to_all_fa, chrom, start, end):\n",
    "        refseq = pyfaidx.Fasta(os.path.join(path_to_all_fa, \"{}.fa\".format(chrom)))\n",
    "        return(str.upper(refseq.get_seq(name = \"{}\".format(chrom), start = start, end = end).seq))\n",
    "\n",
    "path_to_all_fa = \"/home/hieunguyen/resources/hg19\"\n",
    "genome_version = \"hg19\"\n",
    "\n",
    "##### get list of all real cpg for this panel\n",
    "all_cpgdf = dict()\n",
    "for panel_name in final_umi_rundf.Panel_version.unique():\n",
    "    cpgdf = pd.DataFrame()\n",
    "    panel_name = panel_name.replace(\" \", \"_\")\n",
    "    paneldf = pd.read_excel(\"../panel_design/All_panel_designs_20250601.xlsx\", sheet_name=panel_name)\n",
    "    for region in paneldf[genome_version].unique():\n",
    "        region_name = paneldf[paneldf[genome_version] == region][\"Amplicon\"].values[0]\n",
    "        chrom = region.split(\":\")[0].replace(\"Chr\", \"chr\").replace(\"chr\", \"\")\n",
    "        start = int(region.split(\":\")[1].split(\"-\")[0])\n",
    "        end = int(region.split(\":\")[1].split(\"-\")[1])\n",
    "        refseq = pyfaidx.Fasta(os.path.join(path_to_all_fa, \"chr{}.fa\".format(chrom)))\n",
    "        refseq_at_region = str.upper(refseq.get_seq(name = \"chr{}\".format(chrom), start = start, end = end).seq)\n",
    "\n",
    "        all_cpg_in_cluster = [m.start(0) for m in re.finditer(\"CG\", refseq_at_region)]\n",
    "        cpg_coords = [f\"chr{chrom}:{item + start}-{item + start + 1}\" for item in all_cpg_in_cluster]\n",
    "\n",
    "        tmp_cpgdf = pd.DataFrame(data = cpg_coords, columns = [\"CpG\"])\n",
    "        tmp_cpgdf[\"region\"] = region\n",
    "        tmp_cpgdf[\"region_name\"] = region_name\n",
    "        cpgdf = pd.concat([cpgdf, tmp_cpgdf], axis = 0)\n",
    "\n",
    "    cpgdf = cpgdf[[\"region\", \"CpG\", \"region_name\"]]\n",
    "    all_cpgdf[panel_name] = cpgdf.copy()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "502edd17",
   "metadata": {},
   "source": [
    "# PROCESSING COV FILES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e434815b",
   "metadata": {},
   "outputs": [],
   "source": [
    "j = 50\n",
    "\n",
    "panel_name = final_umi_rundf.Panel_version.values[j].replace(\" \", \"_\")\n",
    "run = final_umi_rundf.Run.values[j]\n",
    "mode = \"UMI\"\n",
    "sampleid = final_umi_rundf.SampleID.values[j]\n",
    "if mode == \"UMI\":\n",
    "    input_cov_file = final_umi_rundf.UMI_cov_file.values[j]\n",
    "elif mode == \"ignore_UMI\":\n",
    "    input_cov_file = final_umi_rundf.nonUMI_cov_file.values[j]\n",
    "\n",
    "outputdir = os.path.join(path_to_01_output, \"UMI_cov\", run, panel_name, mode)\n",
    "\n",
    "def generate_cov(input_cov_file, panel_name, sampleid, outputdir):\n",
    "    os.system(f\"mkdir -p {outputdir}\")\n",
    "    if os.path.exists(input_cov_file) == False:\n",
    "        print(f\"Input coverage file {input_cov_file} does not exist, skipping...\")\n",
    "        covdf = pd.DataFrame(columns=[\"chrom\", \"start\", \"end\", \"meth_density\", \"countC\", \"countT\", \"seq\", \"strand\", \"CpG\", \"check_context\", \"amplicon\"])\n",
    "        status = \"File not available\"\n",
    "    else:\n",
    "        print(f\"Processing {input_cov_file} for panel {panel_name}...\")\n",
    "        cpgdf = all_cpgdf[panel_name]\n",
    "        covdf = pd.read_csv(input_cov_file, header = None, sep = \"\\t\")\n",
    "        covdf.columns = [\"chrom\", \"start\", \"end\", \"meth_density\", \"countC\", \"countT\"]\n",
    "        if covdf.shape[0] != 0:\n",
    "            covdf = covdf[covdf[\"chrom\"].isin([\"chrX\", \"chrY\", \"chrMT\"]) == False]\n",
    "            covdf = covdf[covdf[\"chrom\"].str.contains(\"chrUn\") == False]\n",
    "            covdf = covdf[covdf[\"chrom\"].str.contains(\"_\") == False]\n",
    "            covdf[\"seq\"] = covdf[[\"chrom\", \"start\"]].progress_apply(lambda x: get_refseq(path_to_all_fa= path_to_all_fa, \n",
    "                                                                    chrom = x[0], start = x[1], end = x[1] + 1), axis = 1)\n",
    "            covdf[\"strand\"] = covdf[\"seq\"].apply(lambda x: \"+\" if x != \"CG\" else \"-\")\n",
    "            covdf_raw = covdf.copy()\n",
    "            covdf[\"start\"] = covdf[[\"seq\", \"start\"]].apply(lambda x: x[1] + 1 if x[0] != \"CG\" else x[1], axis = 1)\n",
    "\n",
    "            covdf[\"chrom\"] = covdf[\"chrom\"].apply(lambda x: str(x))\n",
    "            covdf[\"CpG\"] = covdf[[\"chrom\", \"start\"]].apply(lambda x: f\"{str(x[0])}:{x[1]}-{x[1] + 1}\", axis = 1)\n",
    "            covdf[\"check_context\"] = covdf[\"CpG\"].apply(lambda x: \"CpG_context\" if x in cpgdf[\"CpG\"].values else \"False\")\n",
    "            covdf[\"amplicon\"] = covdf[\"CpG\"].apply(lambda x: cpgdf[cpgdf[\"CpG\"] == x][\"region_name\"].values[0] if x in cpgdf[\"CpG\"].values else \"NA\")\n",
    "            status = \"Processed\"\n",
    "        else:\n",
    "            print(f\"Coverage file {input_cov_file} is empty, skipping...\")\n",
    "            covdf = pd.DataFrame(columns=[\"chrom\", \"start\", \"end\", \"meth_density\", \"countC\", \"countT\", \"seq\", \"strand\", \"CpG\", \"check_context\", \"amplicon\"])\n",
    "            status = \"Empty file\"\n",
    "        covdf.to_excel(os.path.join(outputdir, f\"{sampleid}.xlsx\"))\n",
    "        return status\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "364538d9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Run', 'SampleID', 'UMI', 'Panel_version', 'Name', 'amplicon_name',\n",
       "       'serial', 'uniqueID', 'filename', 'FASTQ1', 'FASTQ2', 'batch',\n",
       "       'UMI_cov_file', 'nonUMI_cov_file', 'check_UMI', 'check_nonUMI',\n",
       "       'unmapped_bam', 'connor_annotated', 'connor_fastq', 'umi_bam',\n",
       "       'non_umi_bam', 'count_unmapped_bam', 'count_connor_annotated',\n",
       "       'count_connor_fastq', 'count_umi_bam', 'count_non_umi_bam',\n",
       "       'count_UMI_cov_file'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_umi_rundf.columns"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "nextflow_dev",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
