{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e00d9514",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import numpy as np \n",
    "import matplotlib.pyplot as plt \n",
    "import seaborn as sns \n",
    "import os\n",
    "import pathlib\n",
    "import pysam\n",
    "import pyfaidx\n",
    "import warnings\n",
    "import re\n",
    "import argparse\n",
    "import sys\n",
    "import subprocess\n",
    "from tqdm import tqdm \n",
    "\n",
    "tqdm.pandas()\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "main_outptudir = \"/workdir/outdir/pipeline_output\"\n",
    "\n",
    "analysis_output_version = \"20250624\"\n",
    "\n",
    "name = \"Vi\"\n",
    "\n",
    "path_to_main_output = os.path.join(main_outptudir, \"data_analysis\", name, analysis_output_version)\n",
    "\n",
    "\n",
    "rundf = pd.read_excel(\"../panel_design/All_panel_designs_20250601.xlsx\", sheet_name=\"Runs_tracking\")\n",
    "rundf.columns = [\"Run\", \"SampleID\", \"UMI\", \"Panel_version\", \"Name\", \"amplicon_name\", \"serial\"]\n",
    "\n",
    "# umi_rundf = rundf[(rundf[\"UMI\"] == \"Yes\") & (rundf[\"Name\"] == name) ]\n",
    "umi_rundf = rundf[(rundf[\"Name\"] == name) ]\n",
    "\n",
    "collect_all_sample_sheets = [item for item in pathlib.Path(\"../experiments/\").glob(\"SampleSheet_batch_*.csv\")]\n",
    "\n",
    "samplesheet = pd.DataFrame()\n",
    "for file in collect_all_sample_sheets:\n",
    "    df = pd.read_csv(file, sep=\",\")\n",
    "    df[\"batch\"] = file.name.replace(\".csv\", \"\")\n",
    "    samplesheet = pd.concat([samplesheet, df], ignore_index=True)\n",
    "\n",
    "samplesheet.columns = [\"filename\", \"FASTQ1\", \"FASTQ2\", \"batch\"]\n",
    "# samplesheet[\"Run\"] = samplesheet[\"FASTQ1\"].apply(lambda x: x.split(\"/\")[5] if x.split(\"/\")[5] != \"2025\" else x.split(\"/\")[6])\n",
    "samplesheet[\"Run\"] = samplesheet[\"FASTQ1\"].apply(lambda x: [item for item in x.split(\"/\") if item.startswith(\"R\")][0] )\n",
    "samplesheet[\"SampleID\"] = samplesheet[\"filename\"].apply(lambda x: x.split(\"_\")[0].split(\"-\")[1])\n",
    "print(f\"All available runs: {\", \".join(sorted(samplesheet['Run'].unique()))}\")\n",
    "\n",
    "##### R7489\n",
    "selected_run = \"R7489\"\n",
    "UMI = False\n",
    "umt_distance = 0\n",
    "serial = False\n",
    "\n",
    "# ##### R7512\n",
    "# selected_run = \"R7512\"\n",
    "# UMI = False\n",
    "# umt_distance = 0\n",
    "# serial = True\n",
    "\n",
    "##### R7519\n",
    "# selected_run = \"R7519\"\n",
    "# UMI = False\n",
    "# umt_distance = 0\n",
    "# serial = True\n",
    "\n",
    "\n",
    "umi_rundf = umi_rundf[umi_rundf[\"Run\"] == selected_run]\n",
    "samplesheet = samplesheet[samplesheet[\"Run\"] == selected_run]\n",
    "\n",
    "path_to_02_output = os.path.join(path_to_main_output, \"02_output\", selected_run)\n",
    "os.system(f\"mkdir -p {path_to_02_output}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4978caf4",
   "metadata": {},
   "outputs": [],
   "source": [
    "umi_rundf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46eb6be7",
   "metadata": {},
   "outputs": [],
   "source": [
    "samplesheet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ed95853",
   "metadata": {},
   "outputs": [],
   "source": [
    "rerun = True\n",
    "if (os.path.isfile(os.path.join(path_to_02_output, f\"UMI_runs_tracking_{selected_run}.xlsx\")) == False) | (rerun  == True):\n",
    "    umi_rundf[\"uniqueID\"] = umi_rundf[[\"SampleID\", \"Run\"]].apply(lambda x: f\"{x[0]}_{x[1]}\", axis=1)\n",
    "    samplesheet[\"uniqueID\"] = samplesheet[[\"SampleID\", \"Run\"]].apply(lambda x: f\"{x[0]}_{x[1]}\", axis=1)\n",
    "\n",
    "    final_umi_rundf = umi_rundf.merge(samplesheet.drop([\"SampleID\", \"Run\"], axis = 1), left_on = \"uniqueID\", right_on = \"uniqueID\")\n",
    "    final_umi_rundf[\"UMI_cov_file\"] = final_umi_rundf[[\"batch\", \"filename\"]].apply(lambda x: os.path.join(\n",
    "        main_outptudir, x[0], f\"UMT_DISTANCE_{umt_distance}\", \"BISMARK_COV\", f\"{x[1]}.connor_R1_bismark_bt2_pe.bedGraph.gz.bismark.zero.cov\"\n",
    "    ), axis = 1)\n",
    "    final_umi_rundf[\"nonUMI_cov_file\"] = final_umi_rundf[[\"batch\", \"filename\"]].apply(lambda x: os.path.join(\n",
    "        main_outptudir, x[0], f\"UMT_DISTANCE_{umt_distance}\", \"BISMARK_COV\", f\"{x[1]}.CutAdapt_R1_bismark_bt2_pe.bedGraph.gz.bismark.zero.cov\"\n",
    "    ), axis = 1)\n",
    "\n",
    "    final_umi_rundf[\"check_nonUMI\"] = final_umi_rundf[\"nonUMI_cov_file\"].apply(lambda x: os.path.exists(x))\n",
    "\n",
    "    final_umi_rundf[\"unmapped_bam\"] = final_umi_rundf[[\"batch\", \"filename\"]].apply(\n",
    "        lambda x: os.path.join(main_outptudir, x[0], f\"UMT_DISTANCE_{umt_distance}\", \"BISMARK_ALIGNMENT_UNMAPPED_BAM\", f\"{x[1]}_R1.UMIprocessed_bismark_bt2_pe.bam\"), axis = 1\n",
    "    )\n",
    "    if UMI:\n",
    "        final_umi_rundf[\"check_UMI\"] = final_umi_rundf[\"UMI_cov_file\"].apply(lambda x: os.path.exists(x))\n",
    "    \n",
    "        final_umi_rundf[\"connor_annotated\"] = final_umi_rundf[[\"batch\", \"filename\"]].apply(\n",
    "            lambda x: os.path.join(main_outptudir, x[0], f\"UMT_DISTANCE_{umt_distance}\", \"CONNOR_UMI_OUTPUT\", f\"{x[1]}.connor.fully_annotated.bam\"), axis = 1\n",
    "        )\n",
    "        final_umi_rundf[\"connor_fastq\"] = final_umi_rundf[[\"batch\", \"filename\"]].apply(\n",
    "            lambda x: os.path.join(main_outptudir, x[0], f\"UMT_DISTANCE_{umt_distance}\", \"CONNOR_UMI_OUTPUT\", f\"{x[1]}.connor_R1.fastq.gz\"), axis = 1\n",
    "        )\n",
    "        final_umi_rundf[\"umi_bam\"] = final_umi_rundf[[\"batch\", \"filename\"]].apply(\n",
    "            lambda x: os.path.join(main_outptudir, x[0], f\"UMT_DISTANCE_{umt_distance}\", \"BISMARK_ALIGNMENT\", f\"{x[1]}.connor_R1_bismark_bt2_pe.sorted.bam\"), axis = 1\n",
    "        )\n",
    "        available_columns = [\"unmapped_bam\", \"connor_annotated\", \"connor_fastq\", \"umi_bam\", \"non_umi_bam\" , \"UMI_cov_file\"]\n",
    "    \n",
    "    if UMI == False:\n",
    "        available_columns = [\"unmapped_bam\", \"non_umi_bam\"]\n",
    "    final_umi_rundf[\"non_umi_bam\"] = final_umi_rundf[[\"batch\", \"filename\"]].apply(\n",
    "        lambda x: os.path.join(main_outptudir, x[0], f\"UMT_DISTANCE_{umt_distance}\", \"BISMARK_ALIGNMENT\", f\"{x[1]}.CutAdapt_R1_bismark_bt2_pe.sorted.bam\"), axis = 1\n",
    "    )\n",
    "\n",
    "    tqdm.pandas()\n",
    "    for col in available_columns:\n",
    "        print(f\"working on column: {col}\")\n",
    "        def _get_line_count(x):\n",
    "            check_file = os.path.exists(x)\n",
    "            x_raw = x.replace(\".sorted.bam\", \".bam\")\n",
    "            if check_file == False:\n",
    "                if os.path.exists(x_raw) == True: \n",
    "                    # print(f\"sorted bam file is not available, but raw bam file is available, sorting and indexing it now...\")\n",
    "                    os.system(f\"samtools sort -@ 15 {x_raw} -o {x}\")\n",
    "                    os.system(f\"samtools index {x}\")\n",
    "                else:\n",
    "                    # rint(f\"File {x} does not exist, skipping line count for this file.\")\n",
    "                    return 0\n",
    "            if \".bam\" in x:\n",
    "                result = subprocess.run(f\"samtools view {x} -c\", shell=True, capture_output=True, text=True)\n",
    "            elif \".fastq.gz\" in x:\n",
    "                result = subprocess.run(f\"zcat {x} | wc -l\", shell=True, capture_output=True, text=True)\n",
    "            else:\n",
    "                result = subprocess.run(f\"cat {x} | wc -l\", shell=True, capture_output=True, text=True)\n",
    "            return int(result.stdout.strip())\n",
    "            \n",
    "        final_umi_rundf[f\"count_{col}\"] = final_umi_rundf[col].progress_apply(_get_line_count)    \n",
    "    # small size fastq\n",
    "    final_umi_rundf[\"small_FASTQ_size\"] = final_umi_rundf[\"FASTQ1\"].apply(lambda x: \"smallsize\" in x)\n",
    "    final_umi_rundf.groupby(\"small_FASTQ_size\")[\"SampleID\"].count()\n",
    "    final_umi_rundf.to_excel(os.path.join(path_to_02_output, f\"UMI_runs_tracking_{selected_run}.xlsx\"), index=False)\n",
    "else:\n",
    "    print(\"reading in saved processed UMI runs tracking file...\")\n",
    "    final_umi_rundf = pd.read_excel(os.path.join(path_to_02_output, f\"UMI_runs_tracking_{selected_run}.xlsx\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac15dee8",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_umi_rundf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef05a091",
   "metadata": {},
   "source": [
    "# helper functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "088ab731",
   "metadata": {},
   "outputs": [],
   "source": [
    "# panel_name = \"Lung_panel version 1.1\"\n",
    "# cpgdf = pd.DataFrame()\n",
    "# panel_name = panel_name.replace(\" \", \"_\")\n",
    "# paneldf = pd.read_excel(\"../panel_design/All_panel_designs_20250601.xlsx\", sheet_name=panel_name)\n",
    "# paneldf[f\"{genome_version}_org\"] = paneldf[genome_version].values\n",
    "# paneldf[\"lengthF\"] = paneldf[\"Forward Primer Seq\"].apply(lambda x: len(x))\n",
    "# paneldf[\"lengthR\"] = paneldf[\"Reverse Primer Seq\"].apply(lambda x: len(x))\n",
    "\n",
    "def _remove_primer_coords(x, lenF, lenR):\n",
    "    chrom = x.split(\":\")[0]\n",
    "    start = int(x.split(\":\")[1].split(\"-\")[0])\n",
    "    end = int(x.split(\":\")[1].split(\"-\")[1])\n",
    "    start = start + lenF - 1\n",
    "    end = end - lenR + 1\n",
    "    return f\"{chrom}:{start}-{end}\"\n",
    "\n",
    "def get_refseq(path_to_all_fa, chrom, start, end):\n",
    "        refseq = pyfaidx.Fasta(os.path.join(path_to_all_fa, \"{}.fa\".format(chrom)))\n",
    "        return(str.upper(refseq.get_seq(name = \"{}\".format(chrom), start = start, end = end).seq))\n",
    "\n",
    "path_to_all_fa = \"/home/hieunguyen/resources/hg19\"\n",
    "genome_version = \"hg19\"\n",
    "\n",
    "##### get list of all real cpg for this panel\n",
    "all_cpgdf = dict()\n",
    "all_cpgdf_with_primer = dict()\n",
    "all_paneldf = dict()\n",
    "for panel_name in final_umi_rundf.Panel_version.unique():\n",
    "    \n",
    "    panel_name = panel_name.replace(\" \", \"_\")\n",
    "    paneldf = pd.read_excel(\"../panel_design/All_panel_designs_20250601.xlsx\", sheet_name=panel_name)\n",
    "    paneldf[f\"{genome_version}_org\"] = paneldf[genome_version].values\n",
    "    paneldf[\"lengthF\"] = paneldf[\"Forward Primer Seq\"].apply(lambda x: len(x))\n",
    "    paneldf[\"lengthR\"] = paneldf[\"Reverse Primer Seq\"].apply(lambda x: len(x))\n",
    "    paneldf[genome_version] = paneldf[[genome_version, \"lengthF\", \"lengthR\"]].apply(lambda x:_remove_primer_coords(x[0], x[1], x[2]), axis = 1)\n",
    "\n",
    "    def generate_cpgdf(region_col, paneldf):\n",
    "        outputdf = pd.DataFrame()\n",
    "        print(f\"working on panel: {panel_name}, with {len(paneldf[region_col].unique())} regions\")\n",
    "        for region in paneldf[region_col].unique():\n",
    "            region_name = paneldf[paneldf[region_col] == region][\"Amplicon\"].values[0]\n",
    "            chrom = region.split(\":\")[0].replace(\"Chr\", \"chr\").replace(\"chr\", \"\")\n",
    "            start = int(region.split(\":\")[1].split(\"-\")[0])\n",
    "            end = int(region.split(\":\")[1].split(\"-\")[1])\n",
    "            refseq = pyfaidx.Fasta(os.path.join(path_to_all_fa, \"chr{}.fa\".format(chrom)))\n",
    "            refseq_at_region = str.upper(refseq.get_seq(name = \"chr{}\".format(chrom), start = start, end = end).seq)\n",
    "\n",
    "            all_cpg_in_cluster = [m.start(0) for m in re.finditer(\"CG\", refseq_at_region)]\n",
    "            cpg_coords = [f\"chr{chrom}:{item + start}-{item + start + 1}\" for item in all_cpg_in_cluster]\n",
    "\n",
    "            tmp_outputdf = pd.DataFrame(data = cpg_coords, columns = [\"CpG\"])\n",
    "            tmp_outputdf[\"region\"] = region\n",
    "            tmp_outputdf[\"region_name\"] = region_name\n",
    "            outputdf = pd.concat([outputdf, tmp_outputdf], axis = 0)\n",
    "        print(f\"outputdf shape after adding region {region}: {outputdf.shape}\")\n",
    "        return outputdf\n",
    "    \n",
    "    cpgdf = generate_cpgdf(genome_version, paneldf)\n",
    "    cpgdf_with_primer = generate_cpgdf(f\"{genome_version}_org\", paneldf)\n",
    "    \n",
    "    cpgdf = cpgdf[[\"region\", \"CpG\", \"region_name\"]]\n",
    "    all_cpgdf[panel_name] = cpgdf.copy()\n",
    "\n",
    "    cpgdf_with_primer = cpgdf_with_primer[[\"region\", \"CpG\", \"region_name\"]]\n",
    "    all_cpgdf_with_primer[panel_name] = cpgdf_with_primer.copy()\n",
    "\n",
    "    all_paneldf[panel_name] = paneldf.copy()\n",
    "\n",
    "# for i in all_cpgdf.keys():\n",
    "#     tmpdf1 = all_cpgdf[i].copy()\n",
    "#     tmpdf2 = all_cpgdf_with_primer[i].copy()\n",
    "#     assert tmpdf1.shape[0] == tmpdf2.shape[0], f\"Error: {i} cpgdf and cpgdf_with_primer have different number of rows\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "832c8f46",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def generate_cov(input_cov_file, panel_name, sampleid, outputdir):\n",
    "    os.system(f\"mkdir -p {outputdir}\")\n",
    "    if os.path.exists(input_cov_file) == False:\n",
    "        print(f\"Input coverage file {input_cov_file} does not exist, skipping...\")\n",
    "        covdf = pd.DataFrame(columns=[\"chrom\", \"start\", \"end\", \"meth_density\", \"countC\", \"countT\", \"seq\", \"strand\", \"CpG\", \"check_context\", \"amplicon\"])\n",
    "        status = \"File not available\"\n",
    "    else:\n",
    "        print(f\"Processing {input_cov_file} for panel {panel_name}...\")\n",
    "        # get list of CpG which are not in primer regions\n",
    "        cpgdf = all_cpgdf[panel_name]\n",
    "        covdf = pd.read_csv(input_cov_file, header = None, sep = \"\\t\")\n",
    "        covdf.columns = [\"chrom\", \"start\", \"end\", \"meth_density\", \"countC\", \"countT\"]\n",
    "        if covdf.shape[0] != 0:\n",
    "            covdf = covdf[covdf[\"chrom\"].isin([\"chrX\", \"chrY\", \"chrMT\", \"chrM\"]) == False]\n",
    "            covdf = covdf[covdf[\"chrom\"].str.contains(\"chrUn\") == False]\n",
    "            covdf = covdf[covdf[\"chrom\"].str.contains(\"_\") == False]\n",
    "            covdf[\"seq\"] = covdf[[\"chrom\", \"start\"]].progress_apply(lambda x: get_refseq(path_to_all_fa= path_to_all_fa, \n",
    "                                                                    chrom = x[0], start = x[1], end = x[1] + 1), axis = 1)\n",
    "            covdf[\"strand\"] = covdf[\"seq\"].apply(lambda x: \"+\" if x != \"CG\" else \"-\")\n",
    "            covdf_raw = covdf.copy()\n",
    "            covdf[\"start\"] = covdf[[\"seq\", \"start\"]].apply(lambda x: x[1] + 1 if x[0] != \"CG\" else x[1], axis = 1)\n",
    "\n",
    "            covdf[\"chrom\"] = covdf[\"chrom\"].apply(lambda x: str(x))\n",
    "            covdf[\"CpG\"] = covdf[[\"chrom\", \"start\"]].apply(lambda x: f\"{str(x[0])}:{x[1]}-{x[1] + 1}\", axis = 1)\n",
    "            covdf[\"check_context\"] = covdf[\"CpG\"].apply(lambda x: \"CpG_context\" if x in cpgdf[\"CpG\"].values else \"False\")\n",
    "            covdf[\"amplicon\"] = covdf[\"CpG\"].apply(lambda x: cpgdf[cpgdf[\"CpG\"] == x][\"region_name\"].values[0] if x in cpgdf[\"CpG\"].values else \"NA\")\n",
    "            status = \"Processed\"\n",
    "        else:\n",
    "            print(f\"Coverage file {input_cov_file} is empty, skipping...\")\n",
    "            covdf = pd.DataFrame(columns=[\"chrom\", \"start\", \"end\", \"meth_density\", \"countC\", \"countT\", \"seq\", \"strand\", \"CpG\", \"check_context\", \"amplicon\"])\n",
    "            status = \"Empty file\"\n",
    "        covdf.to_excel(os.path.join(outputdir, f\"{sampleid}.xlsx\"))\n",
    "    return status, os.path.join(outputdir, f\"{sampleid}.xlsx\")\n",
    "\n",
    "# mode = \"UMI\"  # or \"ignore_UMI\"\n",
    "if UMI == True:\n",
    "    input_modes = [\"UMI\", \"ignore_UMI\"]\n",
    "else:\n",
    "    input_modes = [\"ignore_UMI\"]\n",
    "for mode in input_modes: \n",
    "    all_status = []\n",
    "    all_finished_path = []\n",
    "    for j in range(final_umi_rundf.shape[0]):\n",
    "        panel_name = final_umi_rundf.Panel_version.values[j].replace(\" \", \"_\")\n",
    "        run = final_umi_rundf.Run.values[j]\n",
    "        sampleid = final_umi_rundf.SampleID.values[j]\n",
    "        if mode == \"UMI\":\n",
    "            input_cov_file = final_umi_rundf.UMI_cov_file.values[j]\n",
    "        elif mode == \"ignore_UMI\":\n",
    "            input_cov_file = final_umi_rundf.nonUMI_cov_file.values[j]\n",
    "\n",
    "        outputdir = os.path.join(path_to_02_output, \"cov\", run, panel_name, mode)\n",
    "        status, output_cov_path = generate_cov(input_cov_file=input_cov_file, panel_name=panel_name, sampleid=sampleid, outputdir=outputdir)\n",
    "        all_status.append(status)\n",
    "        all_finished_path.append(output_cov_path)\n",
    "    final_umi_rundf[f\"{mode}_status\"] = all_status\n",
    "    final_umi_rundf[f\"{mode}_processed_cov\"] = all_finished_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1de5acaf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# run = \"R7288\"\n",
    "# mode = \"UMI\"  # or \"ignore_UMI\"\n",
    "\n",
    "if serial:\n",
    "    for run in final_umi_rundf.Run.unique():\n",
    "    # for run in [\"R7400\"]:\n",
    "        if UMI == True:\n",
    "            input_modes = [\"UMI\", \"ignore_UMI\"]\n",
    "        else:\n",
    "            input_modes = [\"ignore_UMI\"]\n",
    "        for mode in input_modes:\n",
    "            if mode == \"UMI\":\n",
    "                plot_metadata = final_umi_rundf[(final_umi_rundf[\"Run\"] == run) & \n",
    "                                                (final_umi_rundf[\"UMI_status\"] == \"Processed\")].sort_values(by = \"SampleID\")\n",
    "                selected_cov = \"UMI_processed_cov\"\n",
    "            else:\n",
    "                plot_metadata = final_umi_rundf[(final_umi_rundf[\"Run\"] == run) & \n",
    "                                                (final_umi_rundf[\"ignore_UMI_status\"] == \"Processed\")].sort_values(by = \"SampleID\")\n",
    "                selected_cov = \"ignore_UMI_processed_cov\"\n",
    "\n",
    "            run_panels = plot_metadata.Panel_version.unique()\n",
    "\n",
    "            # check how many panels are available for this run\n",
    "            print(f\"Run {run} has {len(plot_metadata.Panel_version.unique())} panel(s), {', '.join(run_panels)}\")\n",
    "\n",
    "            for panel_name in run_panels:\n",
    "                panel_name = panel_name.replace(\" \", \"_\")\n",
    "                cpgdf = all_cpgdf[panel_name].copy()\n",
    "                paneldf = all_paneldf[panel_name].copy()\n",
    "\n",
    "                for amplicon in plot_metadata[plot_metadata[\"amplicon_name\"].isna() == False][\"amplicon_name\"].unique():\n",
    "                    tmp_metadata = plot_metadata[plot_metadata[\"amplicon_name\"] == amplicon].sort_values(by = \"serial\")\n",
    "                    if \",\" in amplicon:\n",
    "                        covdf = pd.DataFrame(data = cpgdf[cpgdf[\"region_name\"].isin(amplicon.replace(\" \", \"\").split(\",\"))][\"CpG\"].values, columns = [\"CpG\"])\n",
    "                    else:\n",
    "                        covdf = pd.DataFrame(data = cpgdf[cpgdf[\"region_name\"] == amplicon][\"CpG\"].values, columns = [\"CpG\"])\n",
    "                        \n",
    "                    os.system(f\"mkdir -p {os.path.join(path_to_02_output, \n",
    "                                                    \"S1_to_S5_in_amplicons\", \n",
    "                                                    run,\n",
    "                                                    panel_name, \n",
    "                                                    mode)}\")\n",
    "                    if tmp_metadata.shape[0] > 1:\n",
    "                        if \",\" in amplicon:\n",
    "                            amplicon = amplicon.replace(\",\", \"-\").replace(\" \", \"\")\n",
    "                        for i in range(tmp_metadata.shape[0]):\n",
    "                            p = tmp_metadata.iloc[i][selected_cov]\n",
    "                            sample  = tmp_metadata.iloc[i][\"serial\"]\n",
    "                            tmp_covdf = pd.read_excel(p)[[\"CpG\", \"meth_density\"]]\n",
    "                            tmp_covdf.columns = [\"CpG\", f\"{amplicon}_{sample}\"]\n",
    "                            covdf = covdf.merge(tmp_covdf, right_on = \"CpG\", left_on = \"CpG\")\n",
    "                        if covdf.shape[0] > 1:\n",
    "                            plt.figure(figsize=(12, 6))\n",
    "                            covdf.set_index(\"CpG\").T.plot.line()\n",
    "                            plt.legend(bbox_to_anchor=(1.05, 1), loc='upper left', borderaxespad=0.)\n",
    "                            plt.xticks(rotation=90)\n",
    "                            plt.tight_layout()\n",
    "                            plt.title(f\"Methylation Density in {amplicon} Amplicon, {mode} \\n Samples: {', '.join(tmp_metadata['serial'].unique())}\")\n",
    "                            plt.savefig(os.path.join(path_to_02_output, \n",
    "                                                    \"S1_to_S5_in_amplicons\", \n",
    "                                                    run,\n",
    "                                                    panel_name, \n",
    "                                                    mode,\n",
    "                                                    f\"{amplicon}_methylation_density.png\"),\n",
    "                                                    dpi=300, \n",
    "                                                    bbox_inches='tight')\n",
    "\n",
    "                            ######\n",
    "                            # corr_plotdf = covdf.set_index(\"CpG\").mean(axis = 0).reset_index()\n",
    "                            # corr_plotdf.columns = [\"CpG\", \"mean_meth_density\"]\n",
    "                            # corr_plotdf[\"true_meth\"] = [0, 5, 10, 50, 100]\n",
    "\n",
    "                            # plt.figure(figsize=(6, 6))\n",
    "                            # sns.scatterplot(data = corr_plotdf, x = \"true_meth\", y = \"mean_meth_density\")\n",
    "                            # corr = corr_plotdf[[\"true_meth\", \"mean_meth_density\"]].corr().iloc[0,1]\n",
    "                            # plt.text(60, 10, f\"Pearson r = {corr:.2f}\", fontsize=14)\n",
    "                            # plt.plot([0, 100], [0, 100], 'r--')\n",
    "                            # plt.xlabel(\"Expected methylation S1-S5 (%)\")\n",
    "                            # plt.ylabel(\"Observed mean methylation density\")\n",
    "                            # plt.title(\"Correlation between true and observed methylation (%)\")\n",
    "                            # plt.tight_layout()\n",
    "                            # plt.savefig(os.path.join(path_to_02_output, \n",
    "                            #                         \"S1_to_S5_in_amplicons\", \n",
    "                            #                         run,\n",
    "                            #                         panel_name, \n",
    "                            #                         mode,\n",
    "                            #                         f\"{amplicon}_correlation.png\"), \n",
    "                            #                         dpi=300, \n",
    "                            #                         bbox_inches='tight')\n",
    "                            # plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8759c23b",
   "metadata": {},
   "outputs": [],
   "source": [
    "###### count on/off target reads\n",
    "def count_read_in_region(bam_path, region, chr_mode = False):\n",
    "    all_reads = []\n",
    "    bamfile = pysam.AlignmentFile(bam_path, \"rb\")\n",
    "    if chr_mode:\n",
    "        region = f\"chr{region}\"\n",
    "    fetched_obj = bamfile.fetch(region = region)\n",
    "    for read in fetched_obj:\n",
    "        all_reads.append(read)\n",
    "    return(len(all_reads))\n",
    "\n",
    "if serial:\n",
    "    for run in final_umi_rundf.Run.unique():\n",
    "        if UMI == True:\n",
    "                input_modes = [\"UMI\", \"ignore_UMI\"]\n",
    "        else:\n",
    "            input_modes = [\"ignore_UMI\"]\n",
    "        for mode in input_modes:\n",
    "            if mode == \"UMI\":\n",
    "                plot_metadata = final_umi_rundf[(final_umi_rundf[\"Run\"] == run) & \n",
    "                                                (final_umi_rundf[\"UMI_status\"] == \"Processed\")].sort_values(by = \"SampleID\")\n",
    "                selected_cov = \"UMI_processed_cov\"\n",
    "                bampath = \"umi_bam\"\n",
    "                total_col = \"count_umi_bam\"\n",
    "            else:\n",
    "                plot_metadata = final_umi_rundf[(final_umi_rundf[\"Run\"] == run) & \n",
    "                                                (final_umi_rundf[\"ignore_UMI_status\"] == \"Processed\")].sort_values(by = \"SampleID\")\n",
    "                selected_cov = \"ignore_UMI_processed_cov\"\n",
    "                bampath = \"non_umi_bam\"\n",
    "                total_col = \"count_non_umi_bam\"\n",
    "\n",
    "            run_panels = plot_metadata.Panel_version.unique()\n",
    "\n",
    "            # check how many panels are available for this run\n",
    "            print(f\"Run {run} has {len(plot_metadata.Panel_version.unique())} panel(s), {', '.join(run_panels)}\")\n",
    "            \n",
    "            for panel_name in run_panels:\n",
    "                tmp_metadata = plot_metadata[plot_metadata[\"Panel_version\"] == panel_name].sort_values(by = [\"amplicon_name\", \"serial\"])\n",
    "                \n",
    "                panel_name = panel_name.replace(\" \", \"_\")\n",
    "                paneldf = all_paneldf[panel_name].copy()\n",
    "                os.system(f\"mkdir -p {os.path.join(path_to_02_output, \n",
    "                                                    \"count_reads_in_regions\", \n",
    "                                                    run,\n",
    "                                                    panel_name, \n",
    "                                                    mode)}\")\n",
    "                for bam_region in tqdm(paneldf[genome_version].unique()):\n",
    "                    region_name = paneldf[paneldf[genome_version] == bam_region][\"Amplicon\"].values[0]\n",
    "                    tmp_metadata[f\"count_{region_name}\"] = tmp_metadata[bampath].apply(lambda x: count_read_in_region(x, bam_region))\n",
    "                    tmp_metadata[f\"pct_{region_name}\"] = tmp_metadata[f\"count_{region_name}\"] / tmp_metadata[total_col] * 100\n",
    "                \n",
    "            tmp_metadata.to_excel(os.path.join(path_to_02_output, \n",
    "                                                \"count_reads_in_regions\", \n",
    "                                                run,\n",
    "                                                panel_name, \n",
    "                                                mode,\n",
    "                                                f\"{run}_{panel_name}.xlsx\"), index = False)\n",
    "            plotdf = tmp_metadata.sort_values(by = \"amplicon_name\")[[item for item in tmp_metadata.columns if \"pct_\" in item] + [\"SampleID\"]].sort_values(by = \"SampleID\").set_index(\"SampleID\")\n",
    "            plt.figure(figsize=(20, 20))\n",
    "            plt.tight_layout()\n",
    "            sns.heatmap(plotdf,cmap=\"YlGnBu\", fmt=\".2f\", annot=True, linewidths=.5, linecolor='black', cbar_kws={\"label\": \"Percentage of reads in region (%)\"})\n",
    "            plt.savefig(os.path.join(path_to_02_output, \n",
    "                                                \"count_reads_in_regions\", \n",
    "                                                run,\n",
    "                                                panel_name, \n",
    "                                                mode,\n",
    "                                                f\"{run}_{panel_name}.png\"), dpi=300, bbox_inches='tight')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c513dda",
   "metadata": {},
   "outputs": [],
   "source": [
    "if serial == True:\n",
    "    for run in final_umi_rundf.Run.unique():\n",
    "    # for run in [\"R7400\"]:\n",
    "        if UMI == True:\n",
    "            input_modes = [\"UMI\", \"ignore_UMI\"]\n",
    "        else:\n",
    "            input_modes = [\"ignore_UMI\"]\n",
    "        for mode in input_modes:\n",
    "            if mode == \"UMI\":\n",
    "                plot_metadata = final_umi_rundf[(final_umi_rundf[\"Run\"] == run) & \n",
    "                                                (final_umi_rundf[\"UMI_status\"] == \"Processed\")].sort_values(by = \"SampleID\")\n",
    "                selected_cov = \"UMI_processed_cov\"\n",
    "                bampath = \"umi_bam\"\n",
    "                total_col = \"count_umi_bam\"\n",
    "            else:\n",
    "                plot_metadata = final_umi_rundf[(final_umi_rundf[\"Run\"] == run) & \n",
    "                                                (final_umi_rundf[\"ignore_UMI_status\"] == \"Processed\")].sort_values(by = \"SampleID\")\n",
    "                selected_cov = \"ignore_UMI_processed_cov\"\n",
    "                bampath = \"non_umi_bam\"\n",
    "                total_col = \"count_non_umi_bam\"\n",
    "\n",
    "            run_panels = plot_metadata.Panel_version.unique()\n",
    "\n",
    "            # check how many panels are available for this run\n",
    "            print(f\"Run {run} has {len(plot_metadata.Panel_version.unique())} panel(s), {', '.join(run_panels)}\")\n",
    "            \n",
    "            for panel_name in run_panels:\n",
    "                metadata = final_umi_rundf[(final_umi_rundf[\"Run\"] == run) & \n",
    "                                        (final_umi_rundf[\"Panel_version\"] == panel_name)\n",
    "                                        & (final_umi_rundf[f\"{mode}_status\"] == \"Processed\")]\n",
    "                panel_name = panel_name.replace(\" \", \"_\")\n",
    "                paneldf = all_paneldf[panel_name].copy()\n",
    "\n",
    "                os.system(f\"mkdir -p {os.path.join(path_to_02_output, \n",
    "                                                    \"methylation_matrix\", \n",
    "                                                    run,\n",
    "                                                    \n",
    "                                                    panel_name, \n",
    "                                                    mode)}\")\n",
    "                if metadata.shape[0] > 1:\n",
    "                    methdf = pd.DataFrame()\n",
    "                    for sampleid in tqdm(metadata.SampleID.unique()):\n",
    "                        covdf = pd.read_excel(metadata[metadata[\"SampleID\"] == sampleid][selected_cov].values[0])\n",
    "                        covdf = covdf[covdf[\"check_context\"] == \"CpG_context\"]\n",
    "                        tmpdf = pd.DataFrame(data = [sampleid], columns = [\"SampleID\"])\n",
    "                        os.system(f\"mkdir -p {os.path.join(path_to_02_output, \"amplicon_cov\", run, panel_name, mode,  sampleid)}\")\n",
    "                        for amplicon in covdf[\"amplicon\"].unique():\n",
    "                            amplicon_covdf = covdf[covdf[\"amplicon\"] == amplicon].copy()\n",
    "                            tmpdf[amplicon] = amplicon_covdf[\"meth_density\"].mean()\n",
    "                            if amplicon_covdf.shape[0] > 0:\n",
    "                                amplicon_covdf.to_excel(os.path.join(path_to_02_output, \"amplicon_cov\", run, panel_name, mode,  sampleid, f\"{sampleid}_{amplicon}_cov.xlsx\"), index = False)\n",
    "                        tmpdf.fillna(0, inplace = True)\n",
    "                        methdf = pd.concat([methdf, tmpdf], axis = 0)\n",
    "                    methdf.sort_values(by = \"SampleID\", inplace = True)\n",
    "                    methdf.to_excel(os.path.join(path_to_02_output, \n",
    "                                                        \"methylation_matrix\", \n",
    "                                                        run,\n",
    "                                                        panel_name, \n",
    "                                                        mode,\n",
    "                                                        \"avg_methylation_density_in_amplicons.xlsx\"), index = False)\n",
    "                    plt.figure(figsize=(35, 20))\n",
    "                    plt.tight_layout()\n",
    "                    sns.heatmap(methdf.set_index(\"SampleID\"),cmap=\"YlGnBu\", fmt=\".2f\", annot=True, linewidths=.5, linecolor='black', cbar_kws={\"label\": \"Average amplicon methylation density (%)\"})\n",
    "                    plt.savefig(os.path.join(path_to_02_output, \n",
    "                                                        \"methylation_matrix\", \n",
    "                                                        run,\n",
    "                                                        panel_name, \n",
    "                                                        mode,\n",
    "                                                        \"avg_methylation_density_in_amplicons.png\"), dpi=300, bbox_inches='tight')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "nextflow_dev",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
