{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e00d9514",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All available runs: R7288, R7297, R7312, R7331, R7332, R7347, R7353, R7373, R7374, R7381, R7393, R7400, R7434, R7477, R7483, R7489, R7512, R7519, R7532, R7547, R7554, R7561, R7568, R7575, R7590, R7611, R7724\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd \n",
    "import numpy as np \n",
    "import matplotlib.pyplot as plt \n",
    "import seaborn as sns \n",
    "import os\n",
    "import pathlib\n",
    "import pysam\n",
    "import pyfaidx\n",
    "import warnings\n",
    "import re\n",
    "import argparse\n",
    "import sys\n",
    "import subprocess\n",
    "from tqdm import tqdm \n",
    "\n",
    "tqdm.pandas()\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "main_outptudir = \"/workdir/outdir/pipeline_output\"\n",
    "\n",
    "analysis_output_version = \"20250729\"\n",
    "\n",
    "name = \"Vi\"\n",
    "# name = \"Truong\"\n",
    "\n",
    "path_to_main_output = os.path.join(main_outptudir, \"data_analysis\", name, analysis_output_version)\n",
    "\n",
    "\n",
    "rundf = pd.read_excel(\"../panel_design/All_panel_designs_20250601.xlsx\", sheet_name=\"Runs_tracking\")\n",
    "rundf.columns = [\"Run\", \"SampleID\", \"UMI\", \"Panel_version\", \"Name\", \"amplicon_name\", \"serial\"]\n",
    "\n",
    "# umi_rundf = rundf[(rundf[\"UMI\"] == \"Yes\") & (rundf[\"Name\"] == name) ]\n",
    "umi_rundf = rundf[(rundf[\"Name\"] == name) ]\n",
    "\n",
    "collect_all_sample_sheets = [item for item in pathlib.Path(\"../experiments/\").glob(\"SampleSheet_batch_*.csv\")]\n",
    "\n",
    "samplesheet = pd.DataFrame()\n",
    "for file in collect_all_sample_sheets:\n",
    "    df = pd.read_csv(file, sep=\",\")\n",
    "    df[\"batch\"] = file.name.replace(\".csv\", \"\")\n",
    "    samplesheet = pd.concat([samplesheet, df], ignore_index=True)\n",
    "\n",
    "samplesheet.columns = [\"filename\", \"FASTQ1\", \"FASTQ2\", \"batch\"]\n",
    "# samplesheet[\"Run\"] = samplesheet[\"FASTQ1\"].apply(lambda x: x.split(\"/\")[5] if x.split(\"/\")[5] != \"2025\" else x.split(\"/\")[6])\n",
    "samplesheet[\"Run\"] = samplesheet[\"FASTQ1\"].apply(lambda x: [item for item in x.split(\"/\") if item.startswith(\"R\")][0] )\n",
    "samplesheet[\"SampleID\"] = samplesheet[\"filename\"].apply(lambda x: x.split(\"_\")[0].split(\"-\")[1])\n",
    "print(f\"All available runs: {\", \".join(sorted(samplesheet['Run'].unique()))}\")\n",
    "\n",
    "##### R7489\n",
    "# selected_run = \"R7489\"\n",
    "# UMI = False\n",
    "# umt_distance = 0\n",
    "# serial = False\n",
    "\n",
    "# ##### R7512\n",
    "# selected_run = \"R7512\"\n",
    "# UMI = False\n",
    "# umt_distance = 0\n",
    "# serial = True\n",
    "\n",
    "##### R7519\n",
    "# selected_run = \"R7519\"\n",
    "# UMI = False\n",
    "# umt_distance = 0\n",
    "# serial = True\n",
    "\n",
    "\n",
    "##### Vi: R7519, R7532, R7554, R7561, R7568, R7611, R7575, R7590\n",
    "##### Truong R7547, R7554, R7561\n",
    "selected_run = \"R7724\"\n",
    "UMI = False\n",
    "umt_distance = 0\n",
    "serial = True\n",
    "\n",
    "if selected_run == \"R7611\":\n",
    "    samplesheet = samplesheet[samplesheet[\"batch\"] == \"SampleSheet_batch_20250721\"]\n",
    "\n",
    "umi_rundf = umi_rundf[umi_rundf[\"Run\"] == selected_run]\n",
    "samplesheet = samplesheet[samplesheet[\"Run\"] == selected_run]\n",
    "\n",
    "path_to_02_output = os.path.join(path_to_main_output, \"02_output\", selected_run)\n",
    "os.system(f\"mkdir -p {path_to_02_output}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "13d52626",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Run</th>\n",
       "      <th>SampleID</th>\n",
       "      <th>UMI</th>\n",
       "      <th>Panel_version</th>\n",
       "      <th>Name</th>\n",
       "      <th>amplicon_name</th>\n",
       "      <th>serial</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>R6801</td>\n",
       "      <td>TM001</td>\n",
       "      <td>No</td>\n",
       "      <td>CRC_panel version 1.1</td>\n",
       "      <td>Vi</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>R6801</td>\n",
       "      <td>TM002</td>\n",
       "      <td>No</td>\n",
       "      <td>CRC_panel version 1.1</td>\n",
       "      <td>Vi</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>R6801</td>\n",
       "      <td>TM005</td>\n",
       "      <td>No</td>\n",
       "      <td>CRC_panel version 1.1</td>\n",
       "      <td>Vi</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>R6801</td>\n",
       "      <td>TM010</td>\n",
       "      <td>No</td>\n",
       "      <td>CRC_panel version 1.1</td>\n",
       "      <td>Vi</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>R6801</td>\n",
       "      <td>TM050</td>\n",
       "      <td>No</td>\n",
       "      <td>CRC_panel version 1.1</td>\n",
       "      <td>Vi</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1073</th>\n",
       "      <td>R7724</td>\n",
       "      <td>TMSTS3V</td>\n",
       "      <td>No</td>\n",
       "      <td>DAPK1</td>\n",
       "      <td>Vi</td>\n",
       "      <td>DAPK1</td>\n",
       "      <td>S3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1074</th>\n",
       "      <td>R7724</td>\n",
       "      <td>TMSTS4V</td>\n",
       "      <td>No</td>\n",
       "      <td>DAPK1</td>\n",
       "      <td>Vi</td>\n",
       "      <td>DAPK1</td>\n",
       "      <td>S4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1075</th>\n",
       "      <td>R7724</td>\n",
       "      <td>TMSTS5V</td>\n",
       "      <td>No</td>\n",
       "      <td>DAPK1</td>\n",
       "      <td>Vi</td>\n",
       "      <td>DAPK1</td>\n",
       "      <td>S5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1076</th>\n",
       "      <td>R7724</td>\n",
       "      <td>TMSTS6V</td>\n",
       "      <td>No</td>\n",
       "      <td>DAPK1</td>\n",
       "      <td>Vi</td>\n",
       "      <td>DAPK1</td>\n",
       "      <td>S6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1077</th>\n",
       "      <td>R7724</td>\n",
       "      <td>TMSTS7V</td>\n",
       "      <td>No</td>\n",
       "      <td>DAPK1</td>\n",
       "      <td>Vi</td>\n",
       "      <td>DAPK1</td>\n",
       "      <td>S7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1078 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Run SampleID UMI          Panel_version Name amplicon_name serial\n",
       "0     R6801    TM001  No  CRC_panel version 1.1   Vi           NaN    NaN\n",
       "1     R6801    TM002  No  CRC_panel version 1.1   Vi           NaN    NaN\n",
       "2     R6801    TM005  No  CRC_panel version 1.1   Vi           NaN    NaN\n",
       "3     R6801    TM010  No  CRC_panel version 1.1   Vi           NaN    NaN\n",
       "4     R6801    TM050  No  CRC_panel version 1.1   Vi           NaN    NaN\n",
       "...     ...      ...  ..                    ...  ...           ...    ...\n",
       "1073  R7724  TMSTS3V  No                  DAPK1   Vi         DAPK1     S3\n",
       "1074  R7724  TMSTS4V  No                  DAPK1   Vi         DAPK1     S4\n",
       "1075  R7724  TMSTS5V  No                  DAPK1   Vi         DAPK1     S5\n",
       "1076  R7724  TMSTS6V  No                  DAPK1   Vi         DAPK1     S6\n",
       "1077  R7724  TMSTS7V  No                  DAPK1   Vi         DAPK1     S7\n",
       "\n",
       "[1078 rows x 7 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rundf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4978caf4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>filename</th>\n",
       "      <th>FASTQ1</th>\n",
       "      <th>FASTQ2</th>\n",
       "      <th>batch</th>\n",
       "      <th>Run</th>\n",
       "      <th>SampleID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>710</th>\n",
       "      <td>10-TMSTS4V_S7561-S7761</td>\n",
       "      <td>/mnt/FASTQ/2025_2/R7724/zatm/10-TMSTS4V_S7561-...</td>\n",
       "      <td>/mnt/FASTQ/2025_2/R7724/zatm/10-TMSTS4V_S7561-...</td>\n",
       "      <td>SampleSheet_batch_20250803</td>\n",
       "      <td>R7724</td>\n",
       "      <td>TMSTS4V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>711</th>\n",
       "      <td>11-TMSTS5V_S7562-S7762</td>\n",
       "      <td>/mnt/FASTQ/2025_2/R7724/zatm/11-TMSTS5V_S7562-...</td>\n",
       "      <td>/mnt/FASTQ/2025_2/R7724/zatm/11-TMSTS5V_S7562-...</td>\n",
       "      <td>SampleSheet_batch_20250803</td>\n",
       "      <td>R7724</td>\n",
       "      <td>TMSTS5V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>712</th>\n",
       "      <td>12-TMSTS6V_S7563-S7763</td>\n",
       "      <td>/mnt/FASTQ/2025_2/R7724/zatm/12-TMSTS6V_S7563-...</td>\n",
       "      <td>/mnt/FASTQ/2025_2/R7724/zatm/12-TMSTS6V_S7563-...</td>\n",
       "      <td>SampleSheet_batch_20250803</td>\n",
       "      <td>R7724</td>\n",
       "      <td>TMSTS6V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>713</th>\n",
       "      <td>13-TMSTS7V_S7564-S7764</td>\n",
       "      <td>/mnt/FASTQ/2025_2/R7724/zatm/13-TMSTS7V_S7564-...</td>\n",
       "      <td>/mnt/FASTQ/2025_2/R7724/zatm/13-TMSTS7V_S7564-...</td>\n",
       "      <td>SampleSheet_batch_20250803</td>\n",
       "      <td>R7724</td>\n",
       "      <td>TMSTS7V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>714</th>\n",
       "      <td>1-TMSTS1P_S7551-S7751</td>\n",
       "      <td>/mnt/FASTQ/2025_2/R7724/zatm/1-TMSTS1P_S7551-S...</td>\n",
       "      <td>/mnt/FASTQ/2025_2/R7724/zatm/1-TMSTS1P_S7551-S...</td>\n",
       "      <td>SampleSheet_batch_20250803</td>\n",
       "      <td>R7724</td>\n",
       "      <td>TMSTS1P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>715</th>\n",
       "      <td>2-TMSTS2P_S7552-S7752</td>\n",
       "      <td>/mnt/FASTQ/2025_2/R7724/zatm/2-TMSTS2P_S7552-S...</td>\n",
       "      <td>/mnt/FASTQ/2025_2/R7724/zatm/2-TMSTS2P_S7552-S...</td>\n",
       "      <td>SampleSheet_batch_20250803</td>\n",
       "      <td>R7724</td>\n",
       "      <td>TMSTS2P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>716</th>\n",
       "      <td>3-TMSTS3P_S7553-S7753</td>\n",
       "      <td>/mnt/FASTQ/2025_2/R7724/zatm/3-TMSTS3P_S7553-S...</td>\n",
       "      <td>/mnt/FASTQ/2025_2/R7724/zatm/3-TMSTS3P_S7553-S...</td>\n",
       "      <td>SampleSheet_batch_20250803</td>\n",
       "      <td>R7724</td>\n",
       "      <td>TMSTS3P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>717</th>\n",
       "      <td>4-TMSTS4P_S7554-S7754</td>\n",
       "      <td>/mnt/FASTQ/2025_2/R7724/zatm/4-TMSTS4P_S7554-S...</td>\n",
       "      <td>/mnt/FASTQ/2025_2/R7724/zatm/4-TMSTS4P_S7554-S...</td>\n",
       "      <td>SampleSheet_batch_20250803</td>\n",
       "      <td>R7724</td>\n",
       "      <td>TMSTS4P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>718</th>\n",
       "      <td>5-TMSTS6P_S7556-S7756</td>\n",
       "      <td>/mnt/FASTQ/2025_2/R7724/zatm/5-TMSTS6P_S7556-S...</td>\n",
       "      <td>/mnt/FASTQ/2025_2/R7724/zatm/5-TMSTS6P_S7556-S...</td>\n",
       "      <td>SampleSheet_batch_20250803</td>\n",
       "      <td>R7724</td>\n",
       "      <td>TMSTS6P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>719</th>\n",
       "      <td>6-TMSTS7P_S7557-S7757</td>\n",
       "      <td>/mnt/FASTQ/2025_2/R7724/zatm/6-TMSTS7P_S7557-S...</td>\n",
       "      <td>/mnt/FASTQ/2025_2/R7724/zatm/6-TMSTS7P_S7557-S...</td>\n",
       "      <td>SampleSheet_batch_20250803</td>\n",
       "      <td>R7724</td>\n",
       "      <td>TMSTS7P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>720</th>\n",
       "      <td>7-TMSTS1V_S7558-S7758</td>\n",
       "      <td>/mnt/FASTQ/2025_2/R7724/zatm/7-TMSTS1V_S7558-S...</td>\n",
       "      <td>/mnt/FASTQ/2025_2/R7724/zatm/7-TMSTS1V_S7558-S...</td>\n",
       "      <td>SampleSheet_batch_20250803</td>\n",
       "      <td>R7724</td>\n",
       "      <td>TMSTS1V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>721</th>\n",
       "      <td>8-TMSTS2V_S7559-S7759</td>\n",
       "      <td>/mnt/FASTQ/2025_2/R7724/zatm/8-TMSTS2V_S7559-S...</td>\n",
       "      <td>/mnt/FASTQ/2025_2/R7724/zatm/8-TMSTS2V_S7559-S...</td>\n",
       "      <td>SampleSheet_batch_20250803</td>\n",
       "      <td>R7724</td>\n",
       "      <td>TMSTS2V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>722</th>\n",
       "      <td>9-TMSTS3V_S7560-S7760</td>\n",
       "      <td>/mnt/FASTQ/2025_2/R7724/zatm/9-TMSTS3V_S7560-S...</td>\n",
       "      <td>/mnt/FASTQ/2025_2/R7724/zatm/9-TMSTS3V_S7560-S...</td>\n",
       "      <td>SampleSheet_batch_20250803</td>\n",
       "      <td>R7724</td>\n",
       "      <td>TMSTS3V</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   filename  \\\n",
       "710  10-TMSTS4V_S7561-S7761   \n",
       "711  11-TMSTS5V_S7562-S7762   \n",
       "712  12-TMSTS6V_S7563-S7763   \n",
       "713  13-TMSTS7V_S7564-S7764   \n",
       "714   1-TMSTS1P_S7551-S7751   \n",
       "715   2-TMSTS2P_S7552-S7752   \n",
       "716   3-TMSTS3P_S7553-S7753   \n",
       "717   4-TMSTS4P_S7554-S7754   \n",
       "718   5-TMSTS6P_S7556-S7756   \n",
       "719   6-TMSTS7P_S7557-S7757   \n",
       "720   7-TMSTS1V_S7558-S7758   \n",
       "721   8-TMSTS2V_S7559-S7759   \n",
       "722   9-TMSTS3V_S7560-S7760   \n",
       "\n",
       "                                                FASTQ1  \\\n",
       "710  /mnt/FASTQ/2025_2/R7724/zatm/10-TMSTS4V_S7561-...   \n",
       "711  /mnt/FASTQ/2025_2/R7724/zatm/11-TMSTS5V_S7562-...   \n",
       "712  /mnt/FASTQ/2025_2/R7724/zatm/12-TMSTS6V_S7563-...   \n",
       "713  /mnt/FASTQ/2025_2/R7724/zatm/13-TMSTS7V_S7564-...   \n",
       "714  /mnt/FASTQ/2025_2/R7724/zatm/1-TMSTS1P_S7551-S...   \n",
       "715  /mnt/FASTQ/2025_2/R7724/zatm/2-TMSTS2P_S7552-S...   \n",
       "716  /mnt/FASTQ/2025_2/R7724/zatm/3-TMSTS3P_S7553-S...   \n",
       "717  /mnt/FASTQ/2025_2/R7724/zatm/4-TMSTS4P_S7554-S...   \n",
       "718  /mnt/FASTQ/2025_2/R7724/zatm/5-TMSTS6P_S7556-S...   \n",
       "719  /mnt/FASTQ/2025_2/R7724/zatm/6-TMSTS7P_S7557-S...   \n",
       "720  /mnt/FASTQ/2025_2/R7724/zatm/7-TMSTS1V_S7558-S...   \n",
       "721  /mnt/FASTQ/2025_2/R7724/zatm/8-TMSTS2V_S7559-S...   \n",
       "722  /mnt/FASTQ/2025_2/R7724/zatm/9-TMSTS3V_S7560-S...   \n",
       "\n",
       "                                                FASTQ2  \\\n",
       "710  /mnt/FASTQ/2025_2/R7724/zatm/10-TMSTS4V_S7561-...   \n",
       "711  /mnt/FASTQ/2025_2/R7724/zatm/11-TMSTS5V_S7562-...   \n",
       "712  /mnt/FASTQ/2025_2/R7724/zatm/12-TMSTS6V_S7563-...   \n",
       "713  /mnt/FASTQ/2025_2/R7724/zatm/13-TMSTS7V_S7564-...   \n",
       "714  /mnt/FASTQ/2025_2/R7724/zatm/1-TMSTS1P_S7551-S...   \n",
       "715  /mnt/FASTQ/2025_2/R7724/zatm/2-TMSTS2P_S7552-S...   \n",
       "716  /mnt/FASTQ/2025_2/R7724/zatm/3-TMSTS3P_S7553-S...   \n",
       "717  /mnt/FASTQ/2025_2/R7724/zatm/4-TMSTS4P_S7554-S...   \n",
       "718  /mnt/FASTQ/2025_2/R7724/zatm/5-TMSTS6P_S7556-S...   \n",
       "719  /mnt/FASTQ/2025_2/R7724/zatm/6-TMSTS7P_S7557-S...   \n",
       "720  /mnt/FASTQ/2025_2/R7724/zatm/7-TMSTS1V_S7558-S...   \n",
       "721  /mnt/FASTQ/2025_2/R7724/zatm/8-TMSTS2V_S7559-S...   \n",
       "722  /mnt/FASTQ/2025_2/R7724/zatm/9-TMSTS3V_S7560-S...   \n",
       "\n",
       "                          batch    Run SampleID  \n",
       "710  SampleSheet_batch_20250803  R7724  TMSTS4V  \n",
       "711  SampleSheet_batch_20250803  R7724  TMSTS5V  \n",
       "712  SampleSheet_batch_20250803  R7724  TMSTS6V  \n",
       "713  SampleSheet_batch_20250803  R7724  TMSTS7V  \n",
       "714  SampleSheet_batch_20250803  R7724  TMSTS1P  \n",
       "715  SampleSheet_batch_20250803  R7724  TMSTS2P  \n",
       "716  SampleSheet_batch_20250803  R7724  TMSTS3P  \n",
       "717  SampleSheet_batch_20250803  R7724  TMSTS4P  \n",
       "718  SampleSheet_batch_20250803  R7724  TMSTS6P  \n",
       "719  SampleSheet_batch_20250803  R7724  TMSTS7P  \n",
       "720  SampleSheet_batch_20250803  R7724  TMSTS1V  \n",
       "721  SampleSheet_batch_20250803  R7724  TMSTS2V  \n",
       "722  SampleSheet_batch_20250803  R7724  TMSTS3V  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samplesheet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ed95853",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "working on column: unmapped_bam\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 13/13 [00:00<00:00, 25479.42it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "working on column: non_umi_bam\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/13 [00:00<?, ?it/s]samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libtinfow.so.6: no version information available (required by samtools)\n",
      "samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libncursesw.so.6: no version information available (required by samtools)\n",
      "samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libncursesw.so.6: no version information available (required by samtools)\n",
      "[bam_sort_core] merging from 0 files and 15 in-memory blocks...\n",
      "samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libtinfow.so.6: no version information available (required by samtools)\n",
      "samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libncursesw.so.6: no version information available (required by samtools)\n",
      "samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libncursesw.so.6: no version information available (required by samtools)\n",
      " 15%|█▌        | 2/13 [00:00<00:05,  2.17it/s]samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libtinfow.so.6: no version information available (required by samtools)\n",
      "samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libncursesw.so.6: no version information available (required by samtools)\n",
      "samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libncursesw.so.6: no version information available (required by samtools)\n",
      "[bam_sort_core] merging from 0 files and 15 in-memory blocks...\n",
      "samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libtinfow.so.6: no version information available (required by samtools)\n",
      "samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libncursesw.so.6: no version information available (required by samtools)\n",
      "samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libncursesw.so.6: no version information available (required by samtools)\n",
      " 23%|██▎       | 3/13 [00:01<00:05,  1.67it/s]samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libtinfow.so.6: no version information available (required by samtools)\n",
      "samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libncursesw.so.6: no version information available (required by samtools)\n",
      "samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libncursesw.so.6: no version information available (required by samtools)\n",
      "[bam_sort_core] merging from 0 files and 15 in-memory blocks...\n",
      "samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libtinfow.so.6: no version information available (required by samtools)\n",
      "samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libncursesw.so.6: no version information available (required by samtools)\n",
      "samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libncursesw.so.6: no version information available (required by samtools)\n",
      " 31%|███       | 4/13 [00:02<00:04,  1.81it/s]samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libtinfow.so.6: no version information available (required by samtools)\n",
      "samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libncursesw.so.6: no version information available (required by samtools)\n",
      "samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libncursesw.so.6: no version information available (required by samtools)\n",
      "[bam_sort_core] merging from 0 files and 15 in-memory blocks...\n",
      "samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libtinfow.so.6: no version information available (required by samtools)\n",
      "samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libncursesw.so.6: no version information available (required by samtools)\n",
      "samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libncursesw.so.6: no version information available (required by samtools)\n",
      " 38%|███▊      | 5/13 [00:04<00:08,  1.09s/it]samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libtinfow.so.6: no version information available (required by samtools)\n",
      "samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libncursesw.so.6: no version information available (required by samtools)\n",
      "samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libncursesw.so.6: no version information available (required by samtools)\n",
      "[bam_sort_core] merging from 0 files and 15 in-memory blocks...\n",
      "samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libtinfow.so.6: no version information available (required by samtools)\n",
      "samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libncursesw.so.6: no version information available (required by samtools)\n",
      "samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libncursesw.so.6: no version information available (required by samtools)\n",
      " 46%|████▌     | 6/13 [00:06<00:11,  1.58s/it]samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libtinfow.so.6: no version information available (required by samtools)\n",
      "samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libncursesw.so.6: no version information available (required by samtools)\n",
      "samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libncursesw.so.6: no version information available (required by samtools)\n",
      "[bam_sort_core] merging from 0 files and 15 in-memory blocks...\n",
      "samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libtinfow.so.6: no version information available (required by samtools)\n",
      "samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libncursesw.so.6: no version information available (required by samtools)\n",
      "samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libncursesw.so.6: no version information available (required by samtools)\n",
      " 54%|█████▍    | 7/13 [00:09<00:10,  1.82s/it]samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libtinfow.so.6: no version information available (required by samtools)\n",
      "samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libncursesw.so.6: no version information available (required by samtools)\n",
      "samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libncursesw.so.6: no version information available (required by samtools)\n",
      "[bam_sort_core] merging from 0 files and 15 in-memory blocks...\n",
      "samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libtinfow.so.6: no version information available (required by samtools)\n",
      "samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libncursesw.so.6: no version information available (required by samtools)\n",
      "samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libncursesw.so.6: no version information available (required by samtools)\n",
      "samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libtinfow.so.6: no version information available (required by samtools)\n",
      "samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libncursesw.so.6: no version information available (required by samtools)\n",
      "samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libncursesw.so.6: no version information available (required by samtools)\n",
      "[bam_sort_core] merging from 0 files and 15 in-memory blocks...\n",
      "samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libtinfow.so.6: no version information available (required by samtools)\n",
      "samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libncursesw.so.6: no version information available (required by samtools)\n",
      "samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libncursesw.so.6: no version information available (required by samtools)\n",
      " 69%|██████▉   | 9/13 [00:09<00:04,  1.01s/it]samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libtinfow.so.6: no version information available (required by samtools)\n",
      "samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libncursesw.so.6: no version information available (required by samtools)\n",
      "samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libncursesw.so.6: no version information available (required by samtools)\n",
      "[bam_sort_core] merging from 0 files and 15 in-memory blocks...\n",
      "samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libtinfow.so.6: no version information available (required by samtools)\n",
      "samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libncursesw.so.6: no version information available (required by samtools)\n",
      "samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libncursesw.so.6: no version information available (required by samtools)\n",
      " 77%|███████▋  | 10/13 [00:09<00:02,  1.13it/s]samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libtinfow.so.6: no version information available (required by samtools)\n",
      "samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libncursesw.so.6: no version information available (required by samtools)\n",
      "samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libncursesw.so.6: no version information available (required by samtools)\n",
      "[bam_sort_core] merging from 0 files and 15 in-memory blocks...\n",
      "samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libtinfow.so.6: no version information available (required by samtools)\n",
      "samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libncursesw.so.6: no version information available (required by samtools)\n",
      "samtools: /home/hieunguyen/miniconda3/envs/nextflow_dev/bin/../lib/libncursesw.so.6: no version information available (required by samtools)\n"
     ]
    }
   ],
   "source": [
    "rerun = True\n",
    "if (os.path.isfile(os.path.join(path_to_02_output, f\"UMI_runs_tracking_{selected_run}.xlsx\")) == False) | (rerun  == True):\n",
    "    umi_rundf[\"uniqueID\"] = umi_rundf[[\"SampleID\", \"Run\"]].apply(lambda x: f\"{x[0]}_{x[1]}\", axis=1)\n",
    "    samplesheet[\"uniqueID\"] = samplesheet[[\"SampleID\", \"Run\"]].apply(lambda x: f\"{x[0]}_{x[1]}\", axis=1)\n",
    "\n",
    "    final_umi_rundf = umi_rundf.merge(samplesheet.drop([\"SampleID\", \"Run\"], axis = 1), left_on = \"uniqueID\", right_on = \"uniqueID\")\n",
    "    final_umi_rundf[\"UMI_cov_file\"] = final_umi_rundf[[\"batch\", \"filename\"]].apply(lambda x: os.path.join(\n",
    "        main_outptudir, x[0], f\"UMT_DISTANCE_{umt_distance}\", \"BISMARK_COV\", f\"{x[1]}.connor_R1_bismark_bt2_pe.bedGraph.gz.bismark.zero.cov\"\n",
    "    ), axis = 1)\n",
    "    final_umi_rundf[\"nonUMI_cov_file\"] = final_umi_rundf[[\"batch\", \"filename\"]].apply(lambda x: os.path.join(\n",
    "        main_outptudir, x[0], f\"UMT_DISTANCE_{umt_distance}\", \"BISMARK_COV\", f\"{x[1]}.CutAdapt_R1_bismark_bt2_pe.bedGraph.gz.bismark.zero.cov\"\n",
    "    ), axis = 1)\n",
    "\n",
    "    final_umi_rundf[\"check_nonUMI\"] = final_umi_rundf[\"nonUMI_cov_file\"].apply(lambda x: os.path.exists(x))\n",
    "\n",
    "    final_umi_rundf[\"unmapped_bam\"] = final_umi_rundf[[\"batch\", \"filename\"]].apply(\n",
    "        lambda x: os.path.join(main_outptudir, x[0], f\"UMT_DISTANCE_{umt_distance}\", \"BISMARK_ALIGNMENT_UNMAPPED_BAM\", f\"{x[1]}_R1.UMIprocessed_bismark_bt2_pe.bam\"), axis = 1\n",
    "    )\n",
    "    if UMI:\n",
    "        final_umi_rundf[\"check_UMI\"] = final_umi_rundf[\"UMI_cov_file\"].apply(lambda x: os.path.exists(x))\n",
    "    \n",
    "        final_umi_rundf[\"connor_annotated\"] = final_umi_rundf[[\"batch\", \"filename\"]].apply(\n",
    "            lambda x: os.path.join(main_outptudir, x[0], f\"UMT_DISTANCE_{umt_distance}\", \"CONNOR_UMI_OUTPUT\", f\"{x[1]}.connor.fully_annotated.bam\"), axis = 1\n",
    "        )\n",
    "        final_umi_rundf[\"connor_fastq\"] = final_umi_rundf[[\"batch\", \"filename\"]].apply(\n",
    "            lambda x: os.path.join(main_outptudir, x[0], f\"UMT_DISTANCE_{umt_distance}\", \"CONNOR_UMI_OUTPUT\", f\"{x[1]}.connor_R1.fastq.gz\"), axis = 1\n",
    "        )\n",
    "        final_umi_rundf[\"umi_bam\"] = final_umi_rundf[[\"batch\", \"filename\"]].apply(\n",
    "            lambda x: os.path.join(main_outptudir, x[0], f\"UMT_DISTANCE_{umt_distance}\", \"BISMARK_ALIGNMENT\", f\"{x[1]}.connor_R1_bismark_bt2_pe.sorted.bam\"), axis = 1\n",
    "        )\n",
    "        available_columns = [\"unmapped_bam\", \"connor_annotated\", \"connor_fastq\", \"umi_bam\", \"non_umi_bam\" , \"UMI_cov_file\"]\n",
    "    \n",
    "    if UMI == False:\n",
    "        available_columns = [\"unmapped_bam\", \"non_umi_bam\"]\n",
    "    final_umi_rundf[\"non_umi_bam\"] = final_umi_rundf[[\"batch\", \"filename\"]].apply(\n",
    "        lambda x: os.path.join(main_outptudir, x[0], f\"UMT_DISTANCE_{umt_distance}\", \"BISMARK_ALIGNMENT\", f\"{x[1]}.CutAdapt_R1_bismark_bt2_pe.sorted.bam\"), axis = 1\n",
    "    )\n",
    "\n",
    "    tqdm.pandas()\n",
    "    for col in available_columns:\n",
    "        print(f\"working on column: {col}\")\n",
    "        def _get_line_count(x):\n",
    "            check_file = os.path.exists(x)\n",
    "            x_raw = x.replace(\".sorted.bam\", \".bam\")\n",
    "            if check_file == False:\n",
    "                if os.path.exists(x_raw) == True: \n",
    "                    # print(f\"sorted bam file is not available, but raw bam file is available, sorting and indexing it now...\")\n",
    "                    os.system(f\"samtools sort -@ 15 {x_raw} -o {x}\")\n",
    "                    os.system(f\"samtools index {x}\")\n",
    "                else:\n",
    "                    # rint(f\"File {x} does not exist, skipping line count for this file.\")\n",
    "                    return 0\n",
    "            if \".bam\" in x:\n",
    "                result = subprocess.run(f\"samtools view {x} -c\", shell=True, capture_output=True, text=True)\n",
    "            elif \".fastq.gz\" in x:\n",
    "                result = subprocess.run(f\"zcat {x} | wc -l\", shell=True, capture_output=True, text=True)\n",
    "            else:\n",
    "                result = subprocess.run(f\"cat {x} | wc -l\", shell=True, capture_output=True, text=True)\n",
    "            return int(result.stdout.strip())\n",
    "            \n",
    "        final_umi_rundf[f\"count_{col}\"] = final_umi_rundf[col].progress_apply(_get_line_count)    \n",
    "    # small size fastq\n",
    "    final_umi_rundf[\"small_FASTQ_size\"] = final_umi_rundf[\"FASTQ1\"].apply(lambda x: \"smallsize\" in x)\n",
    "    final_umi_rundf.groupby(\"small_FASTQ_size\")[\"SampleID\"].count()\n",
    "    final_umi_rundf.to_excel(os.path.join(path_to_02_output, f\"UMI_runs_tracking_{selected_run}.xlsx\"), index=False)\n",
    "else:\n",
    "    print(\"reading in saved processed UMI runs tracking file...\")\n",
    "    final_umi_rundf = pd.read_excel(os.path.join(path_to_02_output, f\"UMI_runs_tracking_{selected_run}.xlsx\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b86eff4",
   "metadata": {},
   "outputs": [],
   "source": [
    "umi_rundf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "532824f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "samplesheet"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef05a091",
   "metadata": {},
   "source": [
    "# helper functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "088ab731",
   "metadata": {},
   "outputs": [],
   "source": [
    "# panel_name = \"Lung_panel version 1.1\"\n",
    "# cpgdf = pd.DataFrame()\n",
    "# panel_name = panel_name.replace(\" \", \"_\")\n",
    "# paneldf = pd.read_excel(\"../panel_design/All_panel_designs_20250601.xlsx\", sheet_name=panel_name)\n",
    "# paneldf[f\"{genome_version}_org\"] = paneldf[genome_version].values\n",
    "# paneldf[\"lengthF\"] = paneldf[\"Forward Primer Seq\"].apply(lambda x: len(x))\n",
    "# paneldf[\"lengthR\"] = paneldf[\"Reverse Primer Seq\"].apply(lambda x: len(x))\n",
    "\n",
    "def _remove_primer_coords(x, lenF, lenR):\n",
    "    chrom = x.split(\":\")[0]\n",
    "    start = int(x.split(\":\")[1].split(\"-\")[0])\n",
    "    end = int(x.split(\":\")[1].split(\"-\")[1])\n",
    "    start = start + lenF - 1\n",
    "    end = end - lenR + 1\n",
    "    return f\"{chrom}:{start}-{end}\"\n",
    "\n",
    "def get_refseq(path_to_all_fa, chrom, start, end):\n",
    "        refseq = pyfaidx.Fasta(os.path.join(path_to_all_fa, \"{}.fa\".format(chrom)))\n",
    "        return(str.upper(refseq.get_seq(name = \"{}\".format(chrom), start = start, end = end).seq))\n",
    "\n",
    "path_to_all_fa = \"/home/hieunguyen/resources/hg19\"\n",
    "genome_version = \"hg19\"\n",
    "\n",
    "##### get list of all real cpg for this panel\n",
    "all_cpgdf = dict()\n",
    "all_cpgdf_with_primer = dict()\n",
    "all_paneldf = dict()\n",
    "for panel_name in tqdm(final_umi_rundf.Panel_version.unique()):\n",
    "    \n",
    "    panel_name = panel_name.replace(\" \", \"_\")\n",
    "    paneldf = pd.read_excel(\"../panel_design/All_panel_designs_20250601.xlsx\", sheet_name=panel_name)\n",
    "    paneldf[f\"{genome_version}_org\"] = paneldf[genome_version].values\n",
    "    paneldf[\"lengthF\"] = paneldf[\"Forward Primer Seq\"].apply(lambda x: len(x))\n",
    "    paneldf[\"lengthR\"] = paneldf[\"Reverse Primer Seq\"].apply(lambda x: len(x))\n",
    "    paneldf[genome_version] = paneldf[[genome_version, \"lengthF\", \"lengthR\"]].apply(lambda x:_remove_primer_coords(x[0], x[1], x[2]), axis = 1)\n",
    "\n",
    "    def generate_cpgdf(region_col, paneldf):\n",
    "        outputdf = pd.DataFrame()\n",
    "        print(f\"working on panel: {panel_name}, with {len(paneldf[region_col].unique())} regions\")\n",
    "        for region in paneldf[region_col].unique():\n",
    "            region_name = paneldf[paneldf[region_col] == region][\"Amplicon\"].values[0]\n",
    "            chrom = region.split(\":\")[0].replace(\"Chr\", \"chr\").replace(\"chr\", \"\")\n",
    "            start = int(region.split(\":\")[1].split(\"-\")[0])\n",
    "            end = int(region.split(\":\")[1].split(\"-\")[1])\n",
    "            refseq = pyfaidx.Fasta(os.path.join(path_to_all_fa, \"chr{}.fa\".format(chrom)))\n",
    "            refseq_at_region = str.upper(refseq.get_seq(name = \"chr{}\".format(chrom), start = start, end = end).seq)\n",
    "\n",
    "            all_cpg_in_cluster = [m.start(0) for m in re.finditer(\"CG\", refseq_at_region)]\n",
    "            cpg_coords = [f\"chr{chrom}:{item + start}-{item + start + 1}\" for item in all_cpg_in_cluster]\n",
    "\n",
    "            tmp_outputdf = pd.DataFrame(data = cpg_coords, columns = [\"CpG\"])\n",
    "            tmp_outputdf[\"region\"] = region\n",
    "            tmp_outputdf[\"region_name\"] = region_name\n",
    "            outputdf = pd.concat([outputdf, tmp_outputdf], axis = 0)\n",
    "        print(f\"outputdf shape after adding region {region}: {outputdf.shape}\")\n",
    "        return outputdf\n",
    "    \n",
    "    cpgdf = generate_cpgdf(genome_version, paneldf)\n",
    "    cpgdf_with_primer = generate_cpgdf(f\"{genome_version}_org\", paneldf)\n",
    "    \n",
    "    cpgdf = cpgdf[[\"region\", \"CpG\", \"region_name\"]]\n",
    "    all_cpgdf[panel_name] = cpgdf.copy()\n",
    "\n",
    "    cpgdf_with_primer = cpgdf_with_primer[[\"region\", \"CpG\", \"region_name\"]]\n",
    "    all_cpgdf_with_primer[panel_name] = cpgdf_with_primer.copy()\n",
    "\n",
    "    all_paneldf[panel_name] = paneldf.copy()\n",
    "\n",
    "# for i in all_cpgdf.keys():\n",
    "#     tmpdf1 = all_cpgdf[i].copy()\n",
    "#     tmpdf2 = all_cpgdf_with_primer[i].copy()\n",
    "#     assert tmpdf1.shape[0] == tmpdf2.shape[0], f\"Error: {i} cpgdf and cpgdf_with_primer have different number of rows\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "881b8dc0",
   "metadata": {},
   "outputs": [],
   "source": [
    "paneldf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "832c8f46",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def generate_cov(input_cov_file, panel_name, sampleid, outputdir):\n",
    "    os.system(f\"mkdir -p {outputdir}\")\n",
    "    if os.path.exists(input_cov_file) == False:\n",
    "        print(f\"Input coverage file {input_cov_file} does not exist, skipping...\")\n",
    "        covdf = pd.DataFrame(columns=[\"chrom\", \"start\", \"end\", \"meth_density\", \"countC\", \"countT\", \"seq\", \"strand\", \"CpG\", \"check_context\", \"amplicon\"])\n",
    "        status = \"File not available\"\n",
    "    else:\n",
    "        print(f\"Processing {input_cov_file} for panel {panel_name}...\")\n",
    "        # get list of CpG which are not in primer regions\n",
    "        cpgdf = all_cpgdf[panel_name]\n",
    "        covdf = pd.read_csv(input_cov_file, header = None, sep = \"\\t\")\n",
    "        covdf.columns = [\"chrom\", \"start\", \"end\", \"meth_density\", \"countC\", \"countT\"]\n",
    "        if covdf.shape[0] != 0:\n",
    "            covdf = covdf[covdf[\"chrom\"].isin([\"chrX\", \"chrY\", \"chrMT\", \"chrM\"]) == False]\n",
    "            covdf = covdf[covdf[\"chrom\"].str.contains(\"chrUn\") == False]\n",
    "            covdf = covdf[covdf[\"chrom\"].str.contains(\"_\") == False]\n",
    "            covdf[\"seq\"] = covdf[[\"chrom\", \"start\"]].progress_apply(lambda x: get_refseq(path_to_all_fa= path_to_all_fa, \n",
    "                                                                    chrom = x[0], start = x[1], end = x[1] + 1), axis = 1)\n",
    "            covdf[\"strand\"] = covdf[\"seq\"].apply(lambda x: \"+\" if x != \"CG\" else \"-\")\n",
    "            covdf_raw = covdf.copy()\n",
    "            covdf[\"start\"] = covdf[[\"seq\", \"start\"]].apply(lambda x: x[1] + 1 if x[0] != \"CG\" else x[1], axis = 1)\n",
    "\n",
    "            covdf[\"chrom\"] = covdf[\"chrom\"].apply(lambda x: str(x))\n",
    "            covdf[\"CpG\"] = covdf[[\"chrom\", \"start\"]].apply(lambda x: f\"{str(x[0])}:{x[1]}-{x[1] + 1}\", axis = 1)\n",
    "            covdf[\"check_context\"] = covdf[\"CpG\"].apply(lambda x: \"CpG_context\" if x in cpgdf[\"CpG\"].values else \"False\")\n",
    "            covdf[\"amplicon\"] = covdf[\"CpG\"].apply(lambda x: cpgdf[cpgdf[\"CpG\"] == x][\"region_name\"].values[0] if x in cpgdf[\"CpG\"].values else \"NA\")\n",
    "            status = \"Processed\"\n",
    "        else:\n",
    "            print(f\"Coverage file {input_cov_file} is empty, skipping...\")\n",
    "            covdf = pd.DataFrame(columns=[\"chrom\", \"start\", \"end\", \"meth_density\", \"countC\", \"countT\", \"seq\", \"strand\", \"CpG\", \"check_context\", \"amplicon\"])\n",
    "            status = \"Empty file\"\n",
    "        covdf.to_excel(os.path.join(outputdir, f\"{sampleid}.xlsx\"))\n",
    "    return status, os.path.join(outputdir, f\"{sampleid}.xlsx\")\n",
    "\n",
    "# mode = \"UMI\"  # or \"ignore_UMI\"\n",
    "if UMI == True:\n",
    "    input_modes = [\"UMI\", \"ignore_UMI\"]\n",
    "else:\n",
    "    input_modes = [\"ignore_UMI\"]\n",
    "for mode in input_modes: \n",
    "    all_status = []\n",
    "    all_finished_path = []\n",
    "    for j in range(final_umi_rundf.shape[0]):\n",
    "        panel_name = final_umi_rundf.Panel_version.values[j].replace(\" \", \"_\")\n",
    "        run = final_umi_rundf.Run.values[j]\n",
    "        sampleid = final_umi_rundf.SampleID.values[j]\n",
    "        if mode == \"UMI\":\n",
    "            input_cov_file = final_umi_rundf.UMI_cov_file.values[j]\n",
    "        elif mode == \"ignore_UMI\":\n",
    "            input_cov_file = final_umi_rundf.nonUMI_cov_file.values[j]\n",
    "\n",
    "        outputdir = os.path.join(path_to_02_output, \"cov\", run, panel_name, mode)\n",
    "        status, output_cov_path = generate_cov(input_cov_file=input_cov_file, panel_name=panel_name, sampleid=sampleid, outputdir=outputdir)\n",
    "        all_status.append(status)\n",
    "        all_finished_path.append(output_cov_path)\n",
    "    final_umi_rundf[f\"{mode}_status\"] = all_status\n",
    "    final_umi_rundf[f\"{mode}_processed_cov\"] = all_finished_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f250959",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_umi_rundf[final_umi_rundf[\"small_FASTQ_size\"] == True]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1de5acaf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# run = \"R7288\"\n",
    "# mode = \"UMI\"  # or \"ignore_UMI\"\n",
    "\n",
    "if serial:\n",
    "    for run in final_umi_rundf.Run.unique():\n",
    "    # for run in [\"R7400\"]:\n",
    "        if UMI == True:\n",
    "            input_modes = [\"UMI\", \"ignore_UMI\"]\n",
    "        else:\n",
    "            input_modes = [\"ignore_UMI\"]\n",
    "        for mode in input_modes:\n",
    "            if mode == \"UMI\":\n",
    "                plot_metadata = final_umi_rundf[(final_umi_rundf[\"Run\"] == run) & \n",
    "                                                (final_umi_rundf[\"UMI_status\"] == \"Processed\")].sort_values(by = \"SampleID\")\n",
    "                selected_cov = \"UMI_processed_cov\"\n",
    "            else:\n",
    "                plot_metadata = final_umi_rundf[(final_umi_rundf[\"Run\"] == run) & \n",
    "                                                (final_umi_rundf[\"ignore_UMI_status\"] == \"Processed\")].sort_values(by = \"SampleID\")\n",
    "                selected_cov = \"ignore_UMI_processed_cov\"\n",
    "\n",
    "            run_panels = plot_metadata.Panel_version.unique()\n",
    "\n",
    "            # check how many panels are available for this run\n",
    "            print(f\"Run {run} has {len(plot_metadata.Panel_version.unique())} panel(s), {', '.join(run_panels)}\")\n",
    "\n",
    "            for panel_name in run_panels:\n",
    "                panel_name = panel_name.replace(\" \", \"_\")\n",
    "                cpgdf = all_cpgdf[panel_name].copy()\n",
    "                paneldf = all_paneldf[panel_name].copy()\n",
    "\n",
    "                for amplicon in plot_metadata[plot_metadata[\"amplicon_name\"].isna() == False][\"amplicon_name\"].unique():\n",
    "                    tmp_metadata = plot_metadata[plot_metadata[\"amplicon_name\"] == amplicon].sort_values(by = \"serial\")\n",
    "                    if \",\" in amplicon:\n",
    "                        covdf = pd.DataFrame(data = cpgdf[cpgdf[\"region_name\"].isin(amplicon.replace(\" \", \"\").split(\",\"))][\"CpG\"].values, columns = [\"CpG\"])\n",
    "                    else:\n",
    "                        covdf = pd.DataFrame(data = cpgdf[cpgdf[\"region_name\"] == amplicon][\"CpG\"].values, columns = [\"CpG\"])\n",
    "                        \n",
    "                    os.system(f\"mkdir -p {os.path.join(path_to_02_output, \n",
    "                                                    \"S1_to_S5_in_amplicons\", \n",
    "                                                    run,\n",
    "                                                    panel_name, \n",
    "                                                    mode)}\")\n",
    "                    if tmp_metadata.shape[0] > 1:\n",
    "                        if \",\" in amplicon:\n",
    "                            amplicon = amplicon.replace(\",\", \"-\").replace(\" \", \"\")\n",
    "                        for i in range(tmp_metadata.shape[0]):\n",
    "                            p = tmp_metadata.iloc[i][selected_cov]\n",
    "                            sample  = tmp_metadata.iloc[i][\"serial\"]\n",
    "                            tmp_covdf = pd.read_excel(p)[[\"CpG\", \"meth_density\"]]\n",
    "                            tmp_covdf.columns = [\"CpG\", f\"{amplicon}_{sample}\"]\n",
    "                            covdf = covdf.merge(tmp_covdf, right_on = \"CpG\", left_on = \"CpG\")\n",
    "                        if covdf.shape[0] > 1:\n",
    "                            plt.figure(figsize=(12, 6))\n",
    "                            covdf.set_index(\"CpG\").T.plot.line()\n",
    "                            plt.legend(bbox_to_anchor=(1.05, 1), loc='upper left', borderaxespad=0.)\n",
    "                            plt.xticks(rotation=90)\n",
    "                            plt.tight_layout()\n",
    "                            plt.title(f\"Methylation Density in {amplicon} Amplicon, {mode} \\n Samples: {', '.join(tmp_metadata['serial'].unique())}\")\n",
    "                            plt.savefig(os.path.join(path_to_02_output, \n",
    "                                                    \"S1_to_S5_in_amplicons\", \n",
    "                                                    run,\n",
    "                                                    panel_name, \n",
    "                                                    mode,\n",
    "                                                    f\"{amplicon}_methylation_density.png\"),\n",
    "                                                    dpi=300, \n",
    "                                                    bbox_inches='tight')\n",
    "\n",
    "                            ######\n",
    "                            # corr_plotdf = covdf.set_index(\"CpG\").mean(axis = 0).reset_index()\n",
    "                            # corr_plotdf.columns = [\"CpG\", \"mean_meth_density\"]\n",
    "                            # corr_plotdf[\"true_meth\"] = [0, 5, 10, 50, 100]\n",
    "\n",
    "                            # plt.figure(figsize=(6, 6))\n",
    "                            # sns.scatterplot(data = corr_plotdf, x = \"true_meth\", y = \"mean_meth_density\")\n",
    "                            # corr = corr_plotdf[[\"true_meth\", \"mean_meth_density\"]].corr().iloc[0,1]\n",
    "                            # plt.text(60, 10, f\"Pearson r = {corr:.2f}\", fontsize=14)\n",
    "                            # plt.plot([0, 100], [0, 100], 'r--')\n",
    "                            # plt.xlabel(\"Expected methylation S1-S5 (%)\")\n",
    "                            # plt.ylabel(\"Observed mean methylation density\")\n",
    "                            # plt.title(\"Correlation between true and observed methylation (%)\")\n",
    "                            # plt.tight_layout()\n",
    "                            # plt.savefig(os.path.join(path_to_02_output, \n",
    "                            #                         \"S1_to_S5_in_amplicons\", \n",
    "                            #                         run,\n",
    "                            #                         panel_name, \n",
    "                            #                         mode,\n",
    "                            #                         f\"{amplicon}_correlation.png\"), \n",
    "                            #                         dpi=300, \n",
    "                            #                         bbox_inches='tight')\n",
    "                            # plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b47b9027",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8759c23b",
   "metadata": {},
   "outputs": [],
   "source": [
    "###### count on/off target reads\n",
    "def count_read_in_region(bam_path, region, chr_mode = False):\n",
    "    all_reads = []\n",
    "    bamfile = pysam.AlignmentFile(bam_path, \"rb\")\n",
    "    if chr_mode:\n",
    "        region = f\"chr{region}\"\n",
    "    fetched_obj = bamfile.fetch(region = region)\n",
    "    for read in fetched_obj:\n",
    "        all_reads.append(read)\n",
    "    return(len(all_reads))\n",
    "\n",
    "# if serial:\n",
    "for run in final_umi_rundf.Run.unique():\n",
    "    if UMI == True:\n",
    "            input_modes = [\"UMI\", \"ignore_UMI\"]\n",
    "    else:\n",
    "        input_modes = [\"ignore_UMI\"]\n",
    "    for mode in input_modes:\n",
    "        if mode == \"UMI\":\n",
    "            plot_metadata = final_umi_rundf[(final_umi_rundf[\"Run\"] == run) & \n",
    "                                            (final_umi_rundf[\"UMI_status\"] == \"Processed\")].sort_values(by = \"SampleID\")\n",
    "            selected_cov = \"UMI_processed_cov\"\n",
    "            bampath = \"umi_bam\"\n",
    "            total_col = \"count_umi_bam\"\n",
    "        else:\n",
    "            plot_metadata = final_umi_rundf[(final_umi_rundf[\"Run\"] == run) & \n",
    "                                            (final_umi_rundf[\"ignore_UMI_status\"] == \"Processed\")].sort_values(by = \"SampleID\")\n",
    "            selected_cov = \"ignore_UMI_processed_cov\"\n",
    "            bampath = \"non_umi_bam\"\n",
    "            total_col = \"count_non_umi_bam\"\n",
    "\n",
    "        run_panels = plot_metadata.Panel_version.unique()\n",
    "\n",
    "        # check how many panels are available for this run\n",
    "        print(f\"Run {run} has {len(plot_metadata.Panel_version.unique())} panel(s), {', '.join(run_panels)}\")\n",
    "        \n",
    "        for panel_name in run_panels:\n",
    "            tmp_metadata = plot_metadata[plot_metadata[\"Panel_version\"] == panel_name].sort_values(by = [\"amplicon_name\", \"serial\"])\n",
    "            \n",
    "            panel_name = panel_name.replace(\" \", \"_\")\n",
    "            paneldf = all_paneldf[panel_name].copy()\n",
    "            os.system(f\"mkdir -p {os.path.join(path_to_02_output, \n",
    "                                                \"count_reads_in_regions\", \n",
    "                                                run,\n",
    "                                                panel_name, \n",
    "                                                mode)}\")\n",
    "            for bam_region in tqdm(paneldf[genome_version].unique()):\n",
    "                region_name = paneldf[paneldf[genome_version] == bam_region][\"Amplicon\"].values[0]\n",
    "                tmp_metadata[f\"count_{region_name}\"] = tmp_metadata[bampath].apply(lambda x: count_read_in_region(x, bam_region))\n",
    "                tmp_metadata[f\"pct_{region_name}\"] = tmp_metadata[f\"count_{region_name}\"] / tmp_metadata[total_col] * 100\n",
    "            \n",
    "        tmp_metadata.to_excel(os.path.join(path_to_02_output, \n",
    "                                            \"count_reads_in_regions\", \n",
    "                                            run,\n",
    "                                            panel_name, \n",
    "                                            mode,\n",
    "                                            f\"{run}_{panel_name}.xlsx\"), index = False)\n",
    "        plotdf = tmp_metadata.sort_values(by = \"amplicon_name\")[[item for item in tmp_metadata.columns if \"pct_\" in item] + [\"SampleID\"]].sort_values(by = \"SampleID\").set_index(\"SampleID\")\n",
    "        plt.figure(figsize=(20, 20))\n",
    "        plt.tight_layout()\n",
    "        sns.heatmap(plotdf,cmap=\"YlGnBu\", fmt=\".2f\", annot=True, linewidths=.5, linecolor='black', cbar_kws={\"label\": \"Percentage of reads in region (%)\"})\n",
    "        plt.savefig(os.path.join(path_to_02_output, \n",
    "                                            \"count_reads_in_regions\", \n",
    "                                            run,\n",
    "                                            panel_name, \n",
    "                                            mode,\n",
    "                                            f\"{run}_{panel_name}.png\"), dpi=300, bbox_inches='tight')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc312987",
   "metadata": {},
   "outputs": [],
   "source": [
    "run_panels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c513dda",
   "metadata": {},
   "outputs": [],
   "source": [
    "# if serial == True:\n",
    "for run in final_umi_rundf.Run.unique():\n",
    "# for run in [\"R7400\"]:\n",
    "    if UMI == True:\n",
    "        input_modes = [\"UMI\", \"ignore_UMI\"]\n",
    "    else:\n",
    "        input_modes = [\"ignore_UMI\"]\n",
    "    for mode in input_modes:\n",
    "        if mode == \"UMI\":\n",
    "            plot_metadata = final_umi_rundf[(final_umi_rundf[\"Run\"] == run) & \n",
    "                                            (final_umi_rundf[\"UMI_status\"] == \"Processed\")].sort_values(by = \"SampleID\")\n",
    "            selected_cov = \"UMI_processed_cov\"\n",
    "            bampath = \"umi_bam\"\n",
    "            total_col = \"count_umi_bam\"\n",
    "        else:\n",
    "            plot_metadata = final_umi_rundf[(final_umi_rundf[\"Run\"] == run) & \n",
    "                                            (final_umi_rundf[\"ignore_UMI_status\"] == \"Processed\")].sort_values(by = \"SampleID\")\n",
    "            selected_cov = \"ignore_UMI_processed_cov\"\n",
    "            bampath = \"non_umi_bam\"\n",
    "            total_col = \"count_non_umi_bam\"\n",
    "\n",
    "        run_panels = plot_metadata.Panel_version.unique()\n",
    "\n",
    "        # check how many panels are available for this run\n",
    "        print(f\"Run {run} has {len(plot_metadata.Panel_version.unique())} panel(s), {', '.join(run_panels)}\")\n",
    "        \n",
    "        for panel_name in run_panels:\n",
    "            metadata = final_umi_rundf[(final_umi_rundf[\"Run\"] == run) & \n",
    "                                    (final_umi_rundf[\"Panel_version\"] == panel_name)\n",
    "                                    & (final_umi_rundf[f\"{mode}_status\"] == \"Processed\")]\n",
    "            panel_name = panel_name.replace(\" \", \"_\")\n",
    "            paneldf = all_paneldf[panel_name].copy()\n",
    "\n",
    "            os.system(f\"mkdir -p {os.path.join(path_to_02_output, \n",
    "                                                \"methylation_matrix\", \n",
    "                                                run,\n",
    "                                                \n",
    "                                                panel_name, \n",
    "                                                mode)}\")\n",
    "            if metadata.shape[0] > 1:\n",
    "                methdf = pd.DataFrame()\n",
    "                for sampleid in tqdm(metadata.SampleID.unique()):\n",
    "                    covdf = pd.read_excel(metadata[metadata[\"SampleID\"] == sampleid][selected_cov].values[0])\n",
    "                    covdf = covdf[covdf[\"check_context\"] == \"CpG_context\"]\n",
    "                    tmpdf = pd.DataFrame(data = [sampleid], columns = [\"SampleID\"])\n",
    "                    os.system(f\"mkdir -p {os.path.join(path_to_02_output, \"amplicon_cov\", run, panel_name, mode,  sampleid)}\")\n",
    "                    for amplicon in covdf[\"amplicon\"].unique():\n",
    "                        amplicon_covdf = covdf[covdf[\"amplicon\"] == amplicon].copy()\n",
    "                        tmpdf[amplicon] = amplicon_covdf[\"meth_density\"].mean()\n",
    "                        if amplicon_covdf.shape[0] > 0:\n",
    "                            amplicon_covdf.to_excel(os.path.join(path_to_02_output, \"amplicon_cov\", run, panel_name, mode,  sampleid, f\"{sampleid}_{amplicon}_cov.xlsx\"), index = False)\n",
    "                    tmpdf.fillna(0, inplace = True)\n",
    "                    methdf = pd.concat([methdf, tmpdf], axis = 0)\n",
    "                methdf.sort_values(by = \"SampleID\", inplace = True)\n",
    "                methdf.to_excel(os.path.join(path_to_02_output, \n",
    "                                                    \"methylation_matrix\", \n",
    "                                                    run,\n",
    "                                                    panel_name, \n",
    "                                                    mode,\n",
    "                                                    \"avg_methylation_density_in_amplicons.xlsx\"), index = False)\n",
    "                plt.figure(figsize=(35, 20))\n",
    "                plt.tight_layout()\n",
    "                sns.heatmap(methdf.set_index(\"SampleID\"),cmap=\"YlGnBu\", fmt=\".2f\", annot=True, linewidths=.5, linecolor='black', cbar_kws={\"label\": \"Average amplicon methylation density (%)\"})\n",
    "                plt.savefig(os.path.join(path_to_02_output, \n",
    "                                                    \"methylation_matrix\", \n",
    "                                                    run,\n",
    "                                                    panel_name, \n",
    "                                                    mode,\n",
    "                                                    \"avg_methylation_density_in_amplicons.png\"), dpi=300, bbox_inches='tight')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "nextflow_dev",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
