{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8572479",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Input directory containing all .cov files for this run: /media/hieunguyen/HNHD01/raw_data/targetMethyl_analysis/target_methylation_R6782_no_dedup_without_non_directional/06_methylation_extract\n",
      "Number of samples in this run R6782: 6\n",
      "                   filename    Run  \\\n",
      "0  5-DML4BCDMR__S7505-S7705  R6782   \n",
      "1  4-DML4BCDMR__S7504-S7704  R6782   \n",
      "2  1-DML4BCDMR__S7501-S7701  R6782   \n",
      "3  6-DML4BCDMR__S7506-S7706  R6782   \n",
      "4  2-DML4BCDMR__S7502-S7702  R6782   \n",
      "5  3-DML4BCDMR__S7503-S7703  R6782   \n",
      "\n",
      "                                                path         mode PIC  \\\n",
      "0  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  directional  Vi   \n",
      "1  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  directional  Vi   \n",
      "2  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  directional  Vi   \n",
      "3  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  directional  Vi   \n",
      "4  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  directional  Vi   \n",
      "5  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  directional  Vi   \n",
      "\n",
      "                                            bam_path  \n",
      "0  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  \n",
      "1  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  \n",
      "2  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  \n",
      "3  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  \n",
      "4  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  \n",
      "5  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  \n",
      "File already exists, skip counting reads in regions. Reading existing data... \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 8/8 [00:00<00:00, 2935.13it/s]\n",
      "100%|██████████| 8/8 [00:00<00:00, 3594.48it/s]\n",
      "100%|██████████| 7/7 [00:00<00:00, 3888.24it/s]\n",
      "100%|██████████| 8/8 [00:00<00:00, 3788.46it/s]\n",
      "100%|██████████| 8/8 [00:00<00:00, 4280.99it/s]\n",
      "100%|██████████| 7/7 [00:00<00:00, 3659.49it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5-DML4BCDMR__S7505-S7705\n",
      "5-DML4BCDMR__S7505-S7705\n",
      "5-DML4BCDMR__S7505-S7705\n",
      "5-DML4BCDMR__S7505-S7705\n",
      "5-DML4BCDMR__S7505-S7705\n",
      "5-DML4BCDMR__S7505-S7705\n",
      "5-DML4BCDMR__S7505-S7705\n",
      "5-DML4BCDMR__S7505-S7705\n",
      "5-DML4BCDMR__S7505-S7705\n",
      "5-DML4BCDMR__S7505-S7705\n",
      "5-DML4BCDMR__S7505-S7705\n",
      "5-DML4BCDMR__S7505-S7705\n",
      "5-DML4BCDMR__S7505-S7705\n",
      "5-DML4BCDMR__S7505-S7705\n",
      "5-DML4BCDMR__S7505-S7705\n",
      "5-DML4BCDMR__S7505-S7705\n",
      "4-DML4BCDMR__S7504-S7704\n",
      "4-DML4BCDMR__S7504-S7704\n",
      "4-DML4BCDMR__S7504-S7704\n",
      "4-DML4BCDMR__S7504-S7704\n",
      "4-DML4BCDMR__S7504-S7704\n",
      "4-DML4BCDMR__S7504-S7704\n",
      "4-DML4BCDMR__S7504-S7704\n",
      "4-DML4BCDMR__S7504-S7704\n",
      "4-DML4BCDMR__S7504-S7704\n",
      "4-DML4BCDMR__S7504-S7704\n",
      "4-DML4BCDMR__S7504-S7704\n",
      "4-DML4BCDMR__S7504-S7704\n",
      "4-DML4BCDMR__S7504-S7704\n",
      "4-DML4BCDMR__S7504-S7704\n",
      "4-DML4BCDMR__S7504-S7704\n",
      "4-DML4BCDMR__S7504-S7704\n",
      "1-DML4BCDMR__S7501-S7701\n",
      "1-DML4BCDMR__S7501-S7701\n",
      "1-DML4BCDMR__S7501-S7701\n",
      "1-DML4BCDMR__S7501-S7701\n",
      "1-DML4BCDMR__S7501-S7701\n",
      "1-DML4BCDMR__S7501-S7701\n",
      "1-DML4BCDMR__S7501-S7701\n",
      "1-DML4BCDMR__S7501-S7701\n",
      "1-DML4BCDMR__S7501-S7701\n",
      "1-DML4BCDMR__S7501-S7701\n",
      "1-DML4BCDMR__S7501-S7701\n",
      "1-DML4BCDMR__S7501-S7701\n",
      "1-DML4BCDMR__S7501-S7701\n",
      "1-DML4BCDMR__S7501-S7701\n",
      "1-DML4BCDMR__S7501-S7701\n",
      "1-DML4BCDMR__S7501-S7701\n",
      "6-DML4BCDMR__S7506-S7706\n",
      "6-DML4BCDMR__S7506-S7706\n",
      "6-DML4BCDMR__S7506-S7706\n",
      "6-DML4BCDMR__S7506-S7706\n",
      "6-DML4BCDMR__S7506-S7706\n",
      "6-DML4BCDMR__S7506-S7706\n",
      "6-DML4BCDMR__S7506-S7706\n",
      "6-DML4BCDMR__S7506-S7706\n",
      "6-DML4BCDMR__S7506-S7706\n",
      "6-DML4BCDMR__S7506-S7706\n",
      "6-DML4BCDMR__S7506-S7706\n",
      "6-DML4BCDMR__S7506-S7706\n",
      "6-DML4BCDMR__S7506-S7706\n",
      "6-DML4BCDMR__S7506-S7706\n",
      "6-DML4BCDMR__S7506-S7706\n",
      "6-DML4BCDMR__S7506-S7706\n",
      "2-DML4BCDMR__S7502-S7702\n",
      "2-DML4BCDMR__S7502-S7702\n",
      "2-DML4BCDMR__S7502-S7702\n",
      "2-DML4BCDMR__S7502-S7702\n",
      "2-DML4BCDMR__S7502-S7702\n",
      "2-DML4BCDMR__S7502-S7702\n",
      "2-DML4BCDMR__S7502-S7702\n",
      "2-DML4BCDMR__S7502-S7702\n",
      "2-DML4BCDMR__S7502-S7702\n",
      "2-DML4BCDMR__S7502-S7702\n",
      "2-DML4BCDMR__S7502-S7702\n",
      "2-DML4BCDMR__S7502-S7702\n",
      "2-DML4BCDMR__S7502-S7702\n",
      "2-DML4BCDMR__S7502-S7702\n",
      "2-DML4BCDMR__S7502-S7702\n",
      "2-DML4BCDMR__S7502-S7702\n",
      "3-DML4BCDMR__S7503-S7703\n",
      "3-DML4BCDMR__S7503-S7703\n",
      "3-DML4BCDMR__S7503-S7703\n",
      "3-DML4BCDMR__S7503-S7703\n",
      "3-DML4BCDMR__S7503-S7703\n",
      "3-DML4BCDMR__S7503-S7703\n",
      "3-DML4BCDMR__S7503-S7703\n",
      "3-DML4BCDMR__S7503-S7703\n",
      "3-DML4BCDMR__S7503-S7703\n",
      "3-DML4BCDMR__S7503-S7703\n",
      "3-DML4BCDMR__S7503-S7703\n",
      "3-DML4BCDMR__S7503-S7703\n",
      "3-DML4BCDMR__S7503-S7703\n",
      "3-DML4BCDMR__S7503-S7703\n",
      "3-DML4BCDMR__S7503-S7703\n",
      "3-DML4BCDMR__S7503-S7703\n",
      "CONTROL114\n",
      "CONTROL114\n",
      "CONTROL114\n",
      "CONTROL114\n",
      "CONTROL114\n",
      "CONTROL114\n",
      "CONTROL114\n",
      "CONTROL114\n",
      "CONTROL114\n",
      "CONTROL114\n",
      "CONTROL114\n",
      "CONTROL114\n",
      "CONTROL114\n",
      "CONTROL114\n",
      "CONTROL114\n",
      "CONTROL114\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd \n",
    "import numpy as np \n",
    "import matplotlib.pyplot as plt \n",
    "import seaborn as sns \n",
    "import os\n",
    "import pathlib\n",
    "import pysam\n",
    "import pyfaidx\n",
    "import warnings\n",
    "import re\n",
    "import argparse\n",
    "import sys\n",
    "from tqdm import tqdm \n",
    "tqdm.pandas()\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "from panel_configs import *\n",
    "from helper_functions import *\n",
    "\n",
    "inputdir = \"/media/hieunguyen/HNHD01/raw_data/targetMethyl_analysis/target_methylation_R6782_no_dedup_without_non_directional/06_methylation_extract\"\n",
    "outputdir = \"/media/hieunguyen/HNHD01/outdir/ampliconSeq/data_analysis/test_get_methylation_data\"\n",
    "path_to_all_fa = \"/media/hieunguyen/GSHD_HN01/storage/resources/hg19\"\n",
    "output_version = \"20250601\"\n",
    "mode = \"directional\"\n",
    "run = \"R6782\"\n",
    "region_version = \"CRC_1.1\"\n",
    "pic = \"Vi\"\n",
    "path_to_control450 = \"/media/hieunguyen/HNSD01/src/ampliconSeq_UMI/CONTROL114_ref_data\"\n",
    "beddir = \"/media/hieunguyen/HNSD01/src/ampliconSeq_UMI/panel_design/beds\"\n",
    "\n",
    "print(f'Input directory containing all .cov files for this run: {inputdir}')\n",
    "\n",
    "path_to_main_output = os.path.join(outputdir,\n",
    "                                \"targetMethyl_analysis\", \n",
    "                                output_version, \n",
    "                                f\"{pic}_output\", \n",
    "                                mode, \n",
    "                                f\"region_version_{region_version}\", \n",
    "                                run)\n",
    "\n",
    "os.system(f\"mkdir -p {path_to_main_output}\")\n",
    "\n",
    "all_cov_files = [item for item in pathlib.Path(inputdir).glob(\"*.cov\")]\n",
    "\n",
    "print(f\"Number of samples in this run {run}: {len(all_cov_files)}\")\n",
    "\n",
    "metadata_dict = {\n",
    "    \"filename\"  : [item.name.split(\".no_deduplicated\")[0] for item in all_cov_files],\n",
    "    # \"Run\" : [str(item).split(\"/\")[6].replace(\"_no_dedup\", \"\").replace(\"target_methylation_\", \"\") for item in all_cov_files],\n",
    "    \"Run\" : [str(item).split(\"/\")[-3] for item in all_cov_files],\n",
    "    \"path\": [str(item) for item in all_cov_files]\n",
    "}\n",
    "\n",
    "metadata = pd.DataFrame.from_dict(metadata_dict, orient=\"columns\")\n",
    "metadata[\"mode\"] = metadata[\"Run\"].apply(lambda x: \"directional\" if \"_without_non_directional\" in x else \"non_directional\")\n",
    "metadata[\"Run\"] = run\n",
    "metadata[\"PIC\"] = pic\n",
    "metadata[\"bam_path\"] = metadata[\"path\"].apply(lambda x: str(x)\\\n",
    "                                                .replace(\"06_methylation_extract\", \"05_sorted_bam\")\\\n",
    "                                                .replace(\".bedGraph.gz.bismark.zero.cov\", \".sorted.bam\"))\n",
    "metadata = metadata[(metadata[\"mode\"] == mode) & (metadata[\"PIC\"] == pic)]\n",
    "print(metadata)\n",
    "metadata.to_excel(os.path.join(path_to_main_output, \"metadata.xlsx\"), index = False)\n",
    "\n",
    "##### regiondf for hg19\n",
    "regiondf = pd.read_csv(os.path.join(beddir, regions[pic][region_version]), sep = \"\\t\", header = None)\n",
    "regiondf.columns = [\"chrom\", \"start\", \"end\", \"amplicon_name\"]\n",
    "regiondf = regiondf[[\"chrom\", \"start\", \"end\", \"amplicon_name\"]]\n",
    "regiondf[\"region_name\"] = regiondf[[\"chrom\", \"start\", \"end\"]].apply(\n",
    "    lambda x: f\"region_{x[0]}_{x[1]}_{x[2]}\", axis = 1\n",
    ")\n",
    "regiondf[\"bam_region\"] = regiondf[[\"chrom\", \"start\", \"end\"]].apply(\n",
    "    lambda x: f\"{x[0].replace('chr', '')}:{x[1]}-{x[2]}\", axis = 1\n",
    ")\n",
    "\n",
    "##### get list of all real cpg for this panel\n",
    "cpgdf = pd.DataFrame()\n",
    "for region in regiondf.region_name.unique():\n",
    "    chrom = region.split(\"_\")[1].replace(\"chr\", \"\")\n",
    "    start = int(region.split(\"_\")[2])\n",
    "    end = int(region.split(\"_\")[3])\n",
    "    refseq = pyfaidx.Fasta(os.path.join(path_to_all_fa, \"chr{}.fa\".format(chrom)))\n",
    "    refseq_at_region = str.upper(refseq.get_seq(name = \"chr{}\".format(chrom), start = start, end = end).seq)\n",
    "\n",
    "    all_cpg_in_cluster = [m.start(0) for m in re.finditer(\"CG\", refseq_at_region)]\n",
    "    cpg_coords = [f\"chr{chrom}:{item + start}-{item + start + 1}\" for item in all_cpg_in_cluster]\n",
    "\n",
    "    tmp_cpgdf = pd.DataFrame(data = cpg_coords, columns = [\"CpG\"])\n",
    "    tmp_cpgdf[\"region\"] = region\n",
    "    cpgdf = pd.concat([cpgdf, tmp_cpgdf], axis = 0)\n",
    "\n",
    "cpgdf = cpgdf[[\"region\", \"CpG\"]]\n",
    "\n",
    "cpgdf.to_excel(os.path.join(path_to_main_output, f\"{pic}_panel_correct_cpgdf.xlsx\"), index = False)\n",
    "\n",
    "###### count on/off target reads\n",
    "metadata[\"num_total_reads\"] = metadata[\"bam_path\"].apply(lambda x: int(pysam.samtools.view(\"-c\", x, catch_stdout=True)))\n",
    "\n",
    "if os.path.isfile(os.path.join(path_to_main_output, f\"read_count_in_region.xlsx\")) == False:\n",
    "    for region_name in regiondf.region_name.unique():\n",
    "        bam_region = regiondf[regiondf[\"region_name\"] == region_name][\"bam_region\"].values[0]\n",
    "        metadata[region_name] = metadata[\"bam_path\"].apply(lambda x: count_read_in_region(x, bam_region))\n",
    "        metadata[f\"pct_{region_name}\"] = metadata[region_name] / metadata[\"num_total_reads\"] * 100\n",
    "        \n",
    "    metadata.to_excel(os.path.join(path_to_main_output, f\"read_count_in_region.xlsx\"), index = False)\n",
    "else:\n",
    "    print(\"File already exists, skip counting reads in regions. Reading existing data... \")\n",
    "    metadata = pd.read_excel(os.path.join(path_to_main_output, f\"read_count_in_region.xlsx\"))\n",
    "\n",
    "##### Process the main COV filesss\n",
    "all_covdf = dict()\n",
    "for run in metadata.Run.unique():\n",
    "    for filename in metadata[metadata[\"Run\"] == run][\"filename\"].unique():\n",
    "            path_to_save_cov = os.path.join(path_to_main_output, filename)\n",
    "            os.system(f\"mkdir -p {path_to_save_cov}\")\n",
    "\n",
    "            covdf = pd.read_csv(metadata[metadata[\"filename\"] == filename][\"path\"].values[0], header = None, sep = \"\\t\")\n",
    "            covdf.columns = [\"chrom\", \"start\", \"end\", \"meth_density\", \"countC\", \"countT\"]\n",
    "            covdf = covdf[covdf[\"chrom\"].isin([\"X\", \"Y\", \"MT\"]) == False]\n",
    "            covdf[\"seq\"] = covdf[[\"chrom\", \"start\"]].progress_apply(lambda x: get_refseq(path_to_all_fa= path_to_all_fa, \n",
    "                                                                    chrom = x[0], start = x[1], end = x[1] + 1), axis = 1)\n",
    "            covdf[\"strand\"] = covdf[\"seq\"].apply(lambda x: \"+\" if x != \"CG\" else \"-\")\n",
    "            covdf_raw = covdf.copy()\n",
    "            covdf_raw.to_excel(f\"{path_to_save_cov}/{filename}_before_modifying_start_coords.xlsx\", index = False)\n",
    "            covdf[\"start\"] = covdf[[\"seq\", \"start\"]].apply(lambda x: x[1] + 1 if x[0] != \"CG\" else x[1], axis = 1)\n",
    "\n",
    "            all_covdf[filename] = covdf\n",
    "            covdf[\"chrom\"] = covdf[\"chrom\"].apply(lambda x: str(x))\n",
    "            strand_name = {\"+\": \"plus\", \"-\": \"minus\"}\n",
    "            for strand in [\"+\", \"-\"]:\n",
    "                for region in regiondf.region_name.unique():\n",
    "                    chrom = str(region.split(\"_\")[1].replace(\"chr\", \"\"))\n",
    "                    start = int(region.split(\"_\")[2])\n",
    "                    end = int(region.split(\"_\")[3])\n",
    "                    save_covdf = covdf[(covdf[\"chrom\"] == chrom) & \n",
    "                            (covdf[\"start\"] >= start) & \n",
    "                            (covdf[\"start\"] <= end) & \n",
    "                            (covdf[\"strand\"] == strand)]\n",
    "                    save_covdf[\"CpG\"] = save_covdf[[\"chrom\", \"start\"]].apply(lambda x: f\"chr{str(x[0])}:{x[1]}-{x[1] + 1}\", axis = 1)\n",
    "                    save_covdf[\"check_context\"] = save_covdf[\"CpG\"].apply(lambda x: \"CpG_context\" if x in cpgdf[\"CpG\"].values else \"False\")\n",
    "                    save_covdf.to_excel(f\"{path_to_save_cov}/{region}_{strand_name[strand]}.xlsx\", index = False)\n",
    "\n",
    "mean_methyldf = pd.DataFrame(data = regiondf.region_name.unique(), columns = [\"region\"])\n",
    "def get_mean_methyl_in_region(region, filename, strand, remove_non_cpg = True):\n",
    "    '''\n",
    "    Calculate average methylation density in a region for a given filename and strand.\n",
    "    If remove_non_cpg is True, it will only keep CpG context sites and remove others.\n",
    "    '''\n",
    "    strand_name = {\"+\": \"plus\", \"-\": \"minus\"}\n",
    "    if filename == \"CONTROL114\":\n",
    "        df = pd.read_excel(f\"{path_to_control450}/region_version_{region_version}/{region}_strand_{strand_name[strand]}.xlsx\")\n",
    "        run = \"R6447\"\n",
    "    else:\n",
    "        run = metadata[metadata[\"filename\"] == filename][\"Run\"].values[0]\n",
    "        df = pd.read_excel(f\"{os.path.join(path_to_main_output, filename)}/{region}_{strand_name[strand]}.xlsx\")\n",
    "    if df.shape[0] == 0:\n",
    "        return \"no data available\"\n",
    "    else:\n",
    "        if remove_non_cpg:\n",
    "            # keep only CpG context Cs\n",
    "            df = df[df[\"check_context\"] == \"CpG_context\"]\n",
    "        df_strand = df[df[\"strand\"] == strand]\n",
    "        if df_strand.shape[0] == 0:\n",
    "            mean_methyl = f\"no read in this region at strand {strand}\"\n",
    "        else:\n",
    "            mean_methyl = df_strand.meth_density.mean()\n",
    "        return mean_methyl\n",
    "\n",
    "for filename in list(all_covdf.keys()) + [\"CONTROL114\"]:\n",
    "    for strand in [\"+\", \"-\"]:\n",
    "        mean_methyldf[f\"{filename}_{strand}\"] = mean_methyldf[\"region\"].apply(lambda x: get_mean_methyl_in_region(x, filename, strand = strand, remove_non_cpg = True))\n",
    "\n",
    "mean_methyldf.to_excel(os.path.join(path_to_main_output, f\"mean_methyl_in_region_compare_CONTROL114.xlsx\"), index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "97ad3d3a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>region</th>\n",
       "      <th>5-DML4BCDMR__S7505-S7705_+</th>\n",
       "      <th>5-DML4BCDMR__S7505-S7705_-</th>\n",
       "      <th>4-DML4BCDMR__S7504-S7704_+</th>\n",
       "      <th>4-DML4BCDMR__S7504-S7704_-</th>\n",
       "      <th>1-DML4BCDMR__S7501-S7701_+</th>\n",
       "      <th>1-DML4BCDMR__S7501-S7701_-</th>\n",
       "      <th>6-DML4BCDMR__S7506-S7706_+</th>\n",
       "      <th>6-DML4BCDMR__S7506-S7706_-</th>\n",
       "      <th>2-DML4BCDMR__S7502-S7702_+</th>\n",
       "      <th>2-DML4BCDMR__S7502-S7702_-</th>\n",
       "      <th>3-DML4BCDMR__S7503-S7703_+</th>\n",
       "      <th>3-DML4BCDMR__S7503-S7703_-</th>\n",
       "      <th>CONTROL114_+</th>\n",
       "      <th>CONTROL114_-</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>region_chr16_22825587_22825729</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>2.083333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>region_chr16_22825770_22825948</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>8.484848</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>region_chr16_22825920_22826053</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>4.886574</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>region_chr16_22826024_22826152</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>5.354811</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>region_chr16_22826130_22826286</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>4.304348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>region_chr17_75369424_75369590</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>region_chr17_75369739_75369887</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>2.792632</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>region_chr17_75369604_75369763</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>no data available</td>\n",
       "      <td>0.892857</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           region 5-DML4BCDMR__S7505-S7705_+  \\\n",
       "0  region_chr16_22825587_22825729          no data available   \n",
       "1  region_chr16_22825770_22825948          no data available   \n",
       "2  region_chr16_22825920_22826053          no data available   \n",
       "3  region_chr16_22826024_22826152          no data available   \n",
       "4  region_chr16_22826130_22826286          no data available   \n",
       "5  region_chr17_75369424_75369590          no data available   \n",
       "6  region_chr17_75369739_75369887          no data available   \n",
       "7  region_chr17_75369604_75369763          no data available   \n",
       "\n",
       "  5-DML4BCDMR__S7505-S7705_- 4-DML4BCDMR__S7504-S7704_+  \\\n",
       "0          no data available          no data available   \n",
       "1          no data available          no data available   \n",
       "2          no data available          no data available   \n",
       "3          no data available          no data available   \n",
       "4          no data available          no data available   \n",
       "5          no data available          no data available   \n",
       "6          no data available          no data available   \n",
       "7          no data available          no data available   \n",
       "\n",
       "  4-DML4BCDMR__S7504-S7704_- 1-DML4BCDMR__S7501-S7701_+  \\\n",
       "0          no data available          no data available   \n",
       "1          no data available          no data available   \n",
       "2          no data available          no data available   \n",
       "3          no data available          no data available   \n",
       "4          no data available          no data available   \n",
       "5          no data available          no data available   \n",
       "6          no data available          no data available   \n",
       "7          no data available          no data available   \n",
       "\n",
       "  1-DML4BCDMR__S7501-S7701_- 6-DML4BCDMR__S7506-S7706_+  \\\n",
       "0          no data available          no data available   \n",
       "1          no data available          no data available   \n",
       "2          no data available          no data available   \n",
       "3          no data available          no data available   \n",
       "4          no data available          no data available   \n",
       "5          no data available          no data available   \n",
       "6          no data available          no data available   \n",
       "7          no data available          no data available   \n",
       "\n",
       "  6-DML4BCDMR__S7506-S7706_- 2-DML4BCDMR__S7502-S7702_+  \\\n",
       "0          no data available          no data available   \n",
       "1          no data available          no data available   \n",
       "2          no data available          no data available   \n",
       "3          no data available          no data available   \n",
       "4          no data available          no data available   \n",
       "5          no data available          no data available   \n",
       "6          no data available          no data available   \n",
       "7          no data available          no data available   \n",
       "\n",
       "  2-DML4BCDMR__S7502-S7702_- 3-DML4BCDMR__S7503-S7703_+  \\\n",
       "0          no data available          no data available   \n",
       "1          no data available          no data available   \n",
       "2          no data available          no data available   \n",
       "3          no data available          no data available   \n",
       "4          no data available          no data available   \n",
       "5          no data available          no data available   \n",
       "6          no data available          no data available   \n",
       "7          no data available          no data available   \n",
       "\n",
       "  3-DML4BCDMR__S7503-S7703_-       CONTROL114_+  CONTROL114_-  \n",
       "0          no data available  no data available      2.083333  \n",
       "1          no data available  no data available      8.484848  \n",
       "2          no data available  no data available      4.886574  \n",
       "3          no data available  no data available      5.354811  \n",
       "4          no data available  no data available      4.304348  \n",
       "5          no data available  no data available      0.000000  \n",
       "6          no data available  no data available      2.792632  \n",
       "7          no data available  no data available      0.892857  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mean_methyldf"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "nextflow_dev",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
